<center><h1>Used Car Price Prediction</h1></center>

# Import Libraries 

In [2]:
from sklearn.model_selection import train_test_split 
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.ensemble import AdaBoostRegressor
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score
from catboost import CatBoostRegressor


from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import GridSearchCV,KFold,RandomizedSearchCV

from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error

from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import KNNImputer,SimpleImputer
import matplotlib.pyplot as plt
import matplotlib

from pymongo import MongoClient

# Read data from MongoDB

In [3]:
from pymongo import MongoClient
  
# Create a pymongo client
client = MongoClient('localhost', 27017)
  
# database instance
db = client['CarData']
  
# collection instance
doc = db['car']
  
# Retrieving first 3 documents using the
# find() and limit() methods 
l1 =[]
for doc1 in doc.find().limit(100000):
    l1.append(doc1)
    #print(doc1)



In [4]:
df = pd.DataFrame(l1)
df

_id          id    year   manufacturer  condition  \
0      62409de88d9a212b80d2832f  7119330988  2012.0           fiat  excellent   
1      62409de88d9a212b80d28330  7109175069  2006.0         toyota       good   
2      62409de88d9a212b80d28331  7109174154  2003.0         toyota       good   
3      62409de88d9a212b80d28332  7109173283  1998.0  mercedes-benz       good   
4      62409de88d9a212b80d28333  7109172375  2007.0          lexus       good   
...                         ...         ...     ...            ...        ...   
99995  62409dea8d9a212b80d407ee  7121385313  2011.0           ford       fair   
99996  62409dea8d9a212b80d407ef  7121360167  2013.0           ford       fair   
99997  62409dea8d9a212b80d407f0  7118198890  2016.0          honda       fair   
99998  62409dea8d9a212b80d407f1  7121333014  2016.0          dodge       fair   
99999  62409dea8d9a212b80d407f2  7121298750  2016.0         nissan       fair   

         cylinders fuel  odometer title_status transmission drive       size  \
0      4 cylinders  gas  142000.0        clean    automatic   fwd  full-size   
1      6 cylinders  gas  159009.0        clean    automatic   4wd  full-size   
2      6 cylinders  gas  151089.0        clean    automatic   4wd  full-size   
3      6 cylinders  gas   88367.0        clean    automatic   4wd  full-size   
4      6 cylinders  gas  169770.0        clean    automatic   4wd  full-size   
...            ...  ...       ...          ...          ...   ...        ...   
99995  5 cylinders  gas  147075.0        clean    automatic   4wd  full-size   
99996  5 cylinders  gas  110875.0        clean    automatic   4wd  full-size   
99997  4 cylinders  gas   40016.0        clean    automatic   fwd  full-size   
99998  5 cylinders  gas   35947.0        clean    automatic   4wd  full-size   
99999  5 cylinders  gas  100888.0        clean    automatic   4wd  full-size   

            type paint_color      lat      long    price  
0        offroad        grey  30.8464  -85.6900   4500.0  
1            van      silver  38.4947  -77.9926   6475.0  
2            SUV         red  38.4947  -77.9926   4995.0  
3            SUV      silver  38.4947  -77.9926   4500.0  
4            SUV       green  38.4947  -77.9926   6995.0  
...          ...         ...      ...       ...      ...  
99995        SUV         red  40.1352  -74.7036   7399.0  
99996     pickup        grey  40.1352  -74.7036  10999.0  
99997      wagon       black  34.0627 -117.6720  12500.0  
99998  hatchback        grey  40.1352  -74.7036  12899.0  
99999      wagon       white  40.1352  -74.7036  12999.0  

[100000 rows x 17 columns]

In [5]:
df = df.drop(["_id"],axis=True)
df = pd.get_dummies(df,drop_first=True)
df.head(5)

id    year  odometer      lat     long   price  \
0  7119330988  2012.0  142000.0  30.8464 -85.6900  4500.0   
1  7109175069  2006.0  159009.0  38.4947 -77.9926  6475.0   
2  7109174154  2003.0  151089.0  38.4947 -77.9926  4995.0   
3  7109173283  1998.0   88367.0  38.4947 -77.9926  4500.0   
4  7109172375  2007.0  169770.0  38.4947 -77.9926  6995.0   

   manufacturer_alfa-romeo  manufacturer_aston-martin  manufacturer_audi  \
0                        0                          0                  0   
1                        0                          0                  0   
2                        0                          0                  0   
3                        0                          0                  0   
4                        0                          0                  0   

   manufacturer_bmw  ...  paint_color_custom  paint_color_green  \
0                 0  ...                   0                  0   
1                 0  ...                   0                  0   
2                 0  ...                   0                  0   
3                 0  ...                   0                  0   
4                 0  ...                   0                  1   

   paint_color_grey  paint_color_orange  paint_color_purple  paint_color_red  \
0                 1                   0                   0                0   
1                 0                   0                   0                0   
2                 0                   0                   0                1   
3                 0                   0                   0                0   
4                 0                   0                   0                0   

   paint_color_silver  paint_color_truck  paint_color_white  \
0                   0                  0                  0   
1                   1                  0                  0   
2                   0                  0                  0   
3                   1                  0                  0   
4                   0                  0                  0   

   paint_color_yellow  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 105 columns]

In [6]:
df.shape

(100000, 105)

In [7]:
print(df.isnull().sum())
print(df.shape)

id                    0
year                  0
odometer              2
lat                   2
long                  0
                     ..
paint_color_red       0
paint_color_silver    0
paint_color_truck     0
paint_color_white     0
paint_color_yellow    0
Length: 105, dtype: int64
(100000, 105)


In [8]:
X=df.drop(["id","price"],axis=True)
y=df["price"]
print(X.shape)
print(y.shape)
print(X.columns)
print(y)

(100000, 103)
(100000,)
Index(['year', 'odometer', 'lat', 'long', 'manufacturer_alfa-romeo',
       'manufacturer_aston-martin', 'manufacturer_audi', 'manufacturer_bmw',
       'manufacturer_buick', 'manufacturer_cadillac',
       ...
       'paint_color_custom', 'paint_color_green', 'paint_color_grey',
       'paint_color_orange', 'paint_color_purple', 'paint_color_red',
       'paint_color_silver', 'paint_color_truck', 'paint_color_white',
       'paint_color_yellow'],
      dtype='object', length=103)
0         4500.0
1         6475.0
2         4995.0
3         4500.0
4         6995.0
          ...   
99995     7399.0
99996    10999.0
99997    12500.0
99998    12899.0
99999    12999.0
Name: price, Length: 100000, dtype: float64


# calculate missing value- using Imputer

In [9]:
imputer = SimpleImputer(strategy='mean')
X =imputer.fit_transform(X) 

In [ ]:
print(X.isnull().sum())
print(X.shape)

# Train Test split

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=2021)

# Linear Regression

In [15]:
#Grid search cv
kfold = KFold(n_splits=5, random_state=2020,shuffle=True)
LR = LinearRegression()
results = cross_val_score(LR, X_train,y_train, cv=kfold)
print(results)
print("R2: %.4f (%.4f)" % (results.mean(), results.std()))

[0.62172227 0.61272732 0.62387275 0.61569704 0.62126806]
R2: 0.6191 (0.0042)


In [18]:
#With Train test split
LR=LinearRegression()
LR.fit(X_train,y_train)
y_pred=LR.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))

r2_score =  0.6193325846505515


In [24]:
cols = pd.DataFrame(X, columns =df.drop(["id","price"],axis=True).columns)

# Ridge Regression

In [26]:
#With Grid Search CV
kfold= KFold(n_splits=5,random_state=2021,shuffle=True)
model = Ridge()
parameters= {'alpha': [0.001,0.1,0.2,0.3,0.4,0.5]}
#fit the grid search
cv= GridSearchCV(model,parameters,scoring='r2',cv=kfold)
cv.fit(X_train,y_train)

#Best Parameters
print(cv.best_params_)
print(cv.best_score_)

y_pred = cv.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=5.54576e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=4.28643e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=4.28746e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=4.26607e-18): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAl

{'alpha': 0.5}
0.6184390929880476
r2_score =  0.6192666442558455


`Ridge Regression
{'alpha': 0.5}
0.6184390929880476`
* r2_score =  0.6192666442558455

# Lasso Regression

In [27]:
#Grid search cv
parameters = dict(alpha=np.linspace(0.001,10))
model = Lasso()
kfold = KFold(n_splits=5, random_state=2021,shuffle=True)
cv = GridSearchCV(model, param_grid=parameters,
                  cv=kfold,scoring='r2')

cv.fit(X_train,y_train)

#Best Parameters
print(cv.best_params_)
print(cv.best_score_)

y_pred = cv.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.376e+12, tolerance: 7.265e+08
  model = cd_fast.enet_coordinate_descent(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.376e+12, tolerance: 7.214e+08
  model = cd_fast.enet_coordinate_descent(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.346e+12, toleranc

{'alpha': 0.20506122448979594}
0.6184460017625509
r2_score =  0.6191765085317995


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.272e+12, tolerance: 9.017e+08
  model = cd_fast.enet_coordinate_descent(


`LASSO Regression
{'alpha': 0.20506122448979594}
0.6184460017625509`
* r2_score =  0.6191765085317995

# Knn Regressor

In [11]:
knn= KNeighborsRegressor()
kfold=KFold(n_splits=5,random_state=2021,shuffle=True)
parameter={"n_neighbors":[1,3,5,7,9,11,13,15]}
cv=GridSearchCV(estimator=knn, param_grid=parameter,
                scoring="r2",cv=kfold,verbose=3)

cv.fit(X_train,y_train)
y_pred = cv.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))


Fitting 5 folds for each of 8 candidates, totalling 40 fits
[CV 1/5] END .....................n_neighbors=1;, score=0.111 total time=  20.1s
[CV 2/5] END .....................n_neighbors=1;, score=0.150 total time=  18.6s
[CV 3/5] END .....................n_neighbors=1;, score=0.087 total time=  17.4s
[CV 4/5] END .....................n_neighbors=1;, score=0.119 total time=  19.0s
[CV 5/5] END .....................n_neighbors=1;, score=0.112 total time=  18.6s
[CV 1/5] END .....................n_neighbors=3;, score=0.293 total time=  24.7s
[CV 2/5] END .....................n_neighbors=3;, score=0.307 total time=  25.6s
[CV 3/5] END .....................n_neighbors=3;, score=0.257 total time=  25.7s
[CV 4/5] END .....................n_neighbors=3;, score=0.279 total time=  25.7s
[CV 5/5] END .....................n_neighbors=3;, score=0.287 total time=  26.0s
[CV 1/5] END .....................n_neighbors=5;, score=0.320 total time=  31.5s
[CV 2/5] END .....................n_neighbors=5;,

`KNN` 
* r2_score =  0.33049468231521995


# Random Forest Regressor

In [11]:
#Grid search cv
model = RandomForestRegressor(random_state=2021)
kfold = KFold(n_splits=5, random_state=2021,shuffle=True)
parameters={'max_features':[6,7,8],
           'n_estimators':[100,150,180],
           'min_samples_leaf':[1,3,5]}

cv=GridSearchCV(estimator=model, param_grid=parameters,
                cv=kfold,verbose=3)

cv.fit(X,y)
print(cv.best_params_)
print(cv.best_score_)

bm_ridge = cv.best_estimator_

Fitting 5 folds for each of 27 candidates, totalling 135 fits
[CV 1/5] END max_features=6, min_samples_leaf=1, n_estimators=100; total time=  47.6s
[CV 2/5] END max_features=6, min_samples_leaf=1, n_estimators=100; total time=  39.6s
[CV 3/5] END max_features=6, min_samples_leaf=1, n_estimators=100; total time=  35.2s
[CV 4/5] END max_features=6, min_samples_leaf=1, n_estimators=100; total time=  38.4s
[CV 5/5] END max_features=6, min_samples_leaf=1, n_estimators=100; total time=  38.7s
[CV 1/5] END max_features=6, min_samples_leaf=1, n_estimators=150; total time=  56.3s
[CV 2/5] END max_features=6, min_samples_leaf=1, n_estimators=150; total time=  57.1s
[CV 3/5] END max_features=6, min_samples_leaf=1, n_estimators=150; total time=  58.4s
[CV 4/5] END max_features=6, min_samples_leaf=1, n_estimators=150; total time=  55.5s
[CV 5/5] END max_features=6, min_samples_leaf=1, n_estimators=150; total time=  53.8s
[CV 1/5] END max_features=6, min_samples_leaf=1, n_estimators=180; total time=

[CV 1/5] END max_features=8, min_samples_leaf=1, n_estimators=150; total time=  47.6s
[CV 2/5] END max_features=8, min_samples_leaf=1, n_estimators=150; total time=  46.9s
[CV 3/5] END max_features=8, min_samples_leaf=1, n_estimators=150; total time=  45.1s
[CV 4/5] END max_features=8, min_samples_leaf=1, n_estimators=150; total time=  44.4s
[CV 5/5] END max_features=8, min_samples_leaf=1, n_estimators=150; total time=  44.3s
[CV 1/5] END max_features=8, min_samples_leaf=1, n_estimators=180; total time= 1.0min
[CV 2/5] END max_features=8, min_samples_leaf=1, n_estimators=180; total time= 1.0min
[CV 3/5] END max_features=8, min_samples_leaf=1, n_estimators=180; total time=  59.9s
[CV 4/5] END max_features=8, min_samples_leaf=1, n_estimators=180; total time=  55.6s
[CV 5/5] END max_features=8, min_samples_leaf=1, n_estimators=180; total time=  49.7s
[CV 1/5] END max_features=8, min_samples_leaf=3, n_estimators=100; total time=  16.3s
[CV 2/5] END max_features=8, min_samples_leaf=3, n_est

`Random Forest
{'max_features': 8, 'min_samples_leaf': 1, 'n_estimators': 180}
0.8184960038334529`
* r2_score =  0.8184960038334529


# Bagging Regressor

In [32]:
#Grid search cv
model = BaggingRegressor(random_state=2021)
kfold= KFold(n_splits=5,random_state=2021,shuffle=True)
parameters= {'n_estimators': [15,20,30,40],'max_features':[5,8,10,15]}
#fit the grid search
cv= GridSearchCV(model,parameters,scoring='r2',cv=kfold,verbose=3)
cv.fit(X_train,y_train)

#Best Parameters
print(cv.best_params_)
print(cv.best_score_)

y_pred = cv.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))

Fitting 5 folds for each of 16 candidates, totalling 80 fits
[CV 1/5] END ...max_features=5, n_estimators=15;, score=0.134 total time=   0.2s
[CV 2/5] END ...max_features=5, n_estimators=15;, score=0.128 total time=   0.2s
[CV 3/5] END ...max_features=5, n_estimators=15;, score=0.126 total time=   0.2s
[CV 4/5] END ...max_features=5, n_estimators=15;, score=0.128 total time=   0.2s
[CV 5/5] END ...max_features=5, n_estimators=15;, score=0.132 total time=   0.2s
[CV 1/5] END ...max_features=5, n_estimators=20;, score=0.184 total time=   0.3s
[CV 2/5] END ...max_features=5, n_estimators=20;, score=0.180 total time=   0.3s
[CV 3/5] END ...max_features=5, n_estimators=20;, score=0.174 total time=   0.3s
[CV 4/5] END ...max_features=5, n_estimators=20;, score=0.179 total time=   0.3s
[CV 5/5] END ...max_features=5, n_estimators=20;, score=0.181 total time=   0.3s
[CV 1/5] END ...max_features=5, n_estimators=30;, score=0.228 total time=   0.6s
[CV 2/5] END ...max_features=5, n_estimators=30;

`Bagging Regresor
{'max_features': 15, 'n_estimators': 40}
0.4218406711762924
r2_score =  0.42881801551338306`

# AdaBoost Regressor

In [33]:
#Grid Search cv
model = AdaBoostRegressor(DecisionTreeRegressor(max_depth=24),random_state=2021)
kfold= KFold(n_splits=5,random_state=2021,shuffle=True)
cv=GridSearchCV(cv=kfold,estimator=model,
                 param_grid={'learning_rate': [0.2, 0.3,0.6],
                'n_estimators': [60, 70, 80, 90,100]},
                scoring='r2',verbose=3)

cv.fit(X_train,y_train)

#Best Parameters
print(cv.best_params_)
print(cv.best_score_)

y_pred = cv.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))

Fitting 5 folds for each of 15 candidates, totalling 75 fits
[CV 1/5] END learning_rate=0.2, n_estimators=60;, score=0.801 total time= 1.0min
[CV 2/5] END learning_rate=0.2, n_estimators=60;, score=0.806 total time= 1.0min
[CV 3/5] END learning_rate=0.2, n_estimators=60;, score=0.774 total time= 1.0min
[CV 4/5] END learning_rate=0.2, n_estimators=60;, score=0.758 total time= 1.1min
[CV 5/5] END learning_rate=0.2, n_estimators=60;, score=0.805 total time= 1.1min
[CV 1/5] END learning_rate=0.2, n_estimators=70;, score=0.802 total time= 1.2min
[CV 2/5] END learning_rate=0.2, n_estimators=70;, score=0.804 total time= 1.6min
[CV 3/5] END learning_rate=0.2, n_estimators=70;, score=0.772 total time= 1.4min
[CV 4/5] END learning_rate=0.2, n_estimators=70;, score=0.764 total time= 1.4min
[CV 5/5] END learning_rate=0.2, n_estimators=70;, score=0.803 total time= 1.3min
[CV 1/5] END learning_rate=0.2, n_estimators=80;, score=0.803 total time= 1.4min
[CV 2/5] END learning_rate=0.2, n_estimators=80;

` AdaBoost Regressor
{'learning_rate': 0.6, 'n_estimators': 100}
0.8005646719615527
r2_score =  0.7992880899482`

# XGBoost

In [12]:
#Grid Search cv
lr_range = [0.001, 0.01, 0.1, 0.2,0.25, 0.3]
n_est_range = [10,20,30,50,100]
md_range = [2,4,6,8,10]

parameters = {'learning_rate':lr_range,'n_estimators':n_est_range,
                'max_depth':md_range}
kfold = KFold(n_splits=5, random_state=2021,shuffle=True)

clf = XGBRegressor(random_state=2021)
cv = GridSearchCV(clf, param_grid=parameters,
                  cv=kfold,scoring='r2')


cv.fit(X_train,y_train)
y_pred = cv.predict(X_test)
print(cv.best_params_)
print(cv.best_score_)

print("r2_score = ",r2_score(y_test, y_pred))

{'learning_rate': 0.3, 'max_depth': 10, 'n_estimators': 100}
0.7968035903222808
r2_score =  0.8079557972846251


`XG Boost
{'learning_rate': 0.3, 'max_depth': 10, 'n_estimators': 100}
0.7968035903222808`
* r2_score =  0.8079557972846251

# CatBoostRegressor

In [13]:
cbr = CatBoostRegressor(learning_rate=0.03,random_state=2021)
cbr.fit(X_train,y_train,verbose=True,plot=True)
y_pred = cbr.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 11138.9268154	total: 162ms	remaining: 2m 41s
1:	learn: 10969.8490010	total: 170ms	remaining: 1m 24s
2:	learn: 10803.8506510	total: 177ms	remaining: 58.8s
3:	learn: 10644.3516473	total: 185ms	remaining: 45.9s
4:	learn: 10494.8814844	total: 192ms	remaining: 38.2s
5:	learn: 10349.7592056	total: 200ms	remaining: 33.1s
6:	learn: 10215.7686744	total: 207ms	remaining: 29.4s
7:	learn: 10080.7606501	total: 215ms	remaining: 26.6s
8:	learn: 9952.4694442	total: 222ms	remaining: 24.4s
9:	learn: 9832.7735688	total: 229ms	remaining: 22.7s
10:	learn: 9716.6466885	total: 236ms	remaining: 21.2s
11:	learn: 9603.8272205	total: 244ms	remaining: 20.1s
12:	learn: 9492.8636224	total: 252ms	remaining: 19.2s
13:	learn: 9390.2066152	total: 264ms	remaining: 18.6s
14:	learn: 9291.3528171	total: 272ms	remaining: 17.9s
15:	learn: 9194.8366383	total: 280ms	remaining: 17.2s
16:	learn: 9102.9943484	total: 288ms	remaining: 16.7s
17:	learn: 9018.0840197	total: 296ms	remaining: 16.1s
18:	learn: 8934.9626165	tota

153:	learn: 6473.5890557	total: 2.53s	remaining: 13.9s
154:	learn: 6470.9107668	total: 2.55s	remaining: 13.9s
155:	learn: 6467.9442693	total: 2.57s	remaining: 13.9s
156:	learn: 6464.4743680	total: 2.58s	remaining: 13.9s
157:	learn: 6461.3814104	total: 2.6s	remaining: 13.9s
158:	learn: 6457.2333247	total: 2.62s	remaining: 13.8s
159:	learn: 6453.5378008	total: 2.63s	remaining: 13.8s
160:	learn: 6450.6968302	total: 2.65s	remaining: 13.8s
161:	learn: 6447.4133181	total: 2.66s	remaining: 13.8s
162:	learn: 6444.4999883	total: 2.68s	remaining: 13.8s
163:	learn: 6440.9639706	total: 2.69s	remaining: 13.7s
164:	learn: 6437.2885582	total: 2.71s	remaining: 13.7s
165:	learn: 6434.3877377	total: 2.73s	remaining: 13.7s
166:	learn: 6431.6886992	total: 2.74s	remaining: 13.7s
167:	learn: 6426.5278068	total: 2.76s	remaining: 13.7s
168:	learn: 6423.9775689	total: 2.77s	remaining: 13.6s
169:	learn: 6420.7438427	total: 2.79s	remaining: 13.6s
170:	learn: 6416.1536950	total: 2.81s	remaining: 13.6s
171:	learn:

313:	learn: 6060.7304192	total: 4.5s	remaining: 9.82s
314:	learn: 6058.5352984	total: 4.5s	remaining: 9.79s
315:	learn: 6056.6835959	total: 4.51s	remaining: 9.77s
316:	learn: 6054.5961840	total: 4.52s	remaining: 9.74s
317:	learn: 6053.5779780	total: 4.53s	remaining: 9.71s
318:	learn: 6051.9332887	total: 4.53s	remaining: 9.68s
319:	learn: 6050.4525769	total: 4.54s	remaining: 9.64s
320:	learn: 6048.8093479	total: 4.54s	remaining: 9.61s
321:	learn: 6047.0756609	total: 4.55s	remaining: 9.58s
322:	learn: 6044.8041613	total: 4.56s	remaining: 9.55s
323:	learn: 6043.5998444	total: 4.56s	remaining: 9.52s
324:	learn: 6041.5889579	total: 4.57s	remaining: 9.49s
325:	learn: 6040.7746373	total: 4.58s	remaining: 9.46s
326:	learn: 6038.7074944	total: 4.58s	remaining: 9.43s
327:	learn: 6037.0957284	total: 4.59s	remaining: 9.4s
328:	learn: 6035.0413150	total: 4.6s	remaining: 9.38s
329:	learn: 6033.6305461	total: 4.6s	remaining: 9.35s
330:	learn: 6032.1589741	total: 4.61s	remaining: 9.32s
331:	learn: 603

475:	learn: 5833.0846865	total: 6.21s	remaining: 6.83s
476:	learn: 5832.4123189	total: 6.22s	remaining: 6.82s
477:	learn: 5831.0399223	total: 6.24s	remaining: 6.81s
478:	learn: 5829.5642733	total: 6.25s	remaining: 6.8s
479:	learn: 5827.6382581	total: 6.27s	remaining: 6.79s
480:	learn: 5826.5884816	total: 6.28s	remaining: 6.78s
481:	learn: 5825.1974263	total: 6.29s	remaining: 6.76s
482:	learn: 5823.5630417	total: 6.31s	remaining: 6.75s
483:	learn: 5822.1747243	total: 6.32s	remaining: 6.74s
484:	learn: 5821.2495720	total: 6.34s	remaining: 6.73s
485:	learn: 5820.2011587	total: 6.35s	remaining: 6.72s
486:	learn: 5818.6841655	total: 6.37s	remaining: 6.71s
487:	learn: 5817.5714484	total: 6.38s	remaining: 6.69s
488:	learn: 5816.4402156	total: 6.4s	remaining: 6.68s
489:	learn: 5815.4550556	total: 6.41s	remaining: 6.67s
490:	learn: 5814.5589120	total: 6.42s	remaining: 6.66s
491:	learn: 5814.0135811	total: 6.44s	remaining: 6.65s
492:	learn: 5812.7779348	total: 6.45s	remaining: 6.64s
493:	learn: 

632:	learn: 5671.1318219	total: 8.62s	remaining: 5s
633:	learn: 5670.0729953	total: 8.63s	remaining: 4.98s
634:	learn: 5668.5246294	total: 8.64s	remaining: 4.96s
635:	learn: 5667.6926661	total: 8.64s	remaining: 4.95s
636:	learn: 5666.5464899	total: 8.65s	remaining: 4.93s
637:	learn: 5666.1310905	total: 8.66s	remaining: 4.91s
638:	learn: 5665.6038604	total: 8.67s	remaining: 4.9s
639:	learn: 5664.8799941	total: 8.68s	remaining: 4.88s
640:	learn: 5664.0832543	total: 8.68s	remaining: 4.86s
641:	learn: 5663.2263814	total: 8.69s	remaining: 4.85s
642:	learn: 5662.7137516	total: 8.7s	remaining: 4.83s
643:	learn: 5661.7804578	total: 8.71s	remaining: 4.81s
644:	learn: 5660.6731413	total: 8.71s	remaining: 4.8s
645:	learn: 5659.6553049	total: 8.72s	remaining: 4.78s
646:	learn: 5658.7086706	total: 8.73s	remaining: 4.76s
647:	learn: 5658.2146354	total: 8.74s	remaining: 4.75s
648:	learn: 5657.5844185	total: 8.75s	remaining: 4.73s
649:	learn: 5656.8450238	total: 8.76s	remaining: 4.71s
650:	learn: 5656

789:	learn: 5544.9434721	total: 10.4s	remaining: 2.77s
790:	learn: 5544.3895347	total: 10.4s	remaining: 2.76s
791:	learn: 5543.4890039	total: 10.5s	remaining: 2.75s
792:	learn: 5542.5267248	total: 10.5s	remaining: 2.74s
793:	learn: 5541.9127996	total: 10.5s	remaining: 2.72s
794:	learn: 5541.1436531	total: 10.5s	remaining: 2.71s
795:	learn: 5540.5496183	total: 10.5s	remaining: 2.7s
796:	learn: 5540.0147343	total: 10.6s	remaining: 2.69s
797:	learn: 5539.5235173	total: 10.6s	remaining: 2.68s
798:	learn: 5538.3508329	total: 10.6s	remaining: 2.66s
799:	learn: 5537.4677080	total: 10.6s	remaining: 2.65s
800:	learn: 5536.8769908	total: 10.6s	remaining: 2.64s
801:	learn: 5535.7520747	total: 10.6s	remaining: 2.63s
802:	learn: 5535.6347856	total: 10.7s	remaining: 2.62s
803:	learn: 5534.9085507	total: 10.7s	remaining: 2.6s
804:	learn: 5534.6662428	total: 10.7s	remaining: 2.59s
805:	learn: 5534.1517276	total: 10.7s	remaining: 2.58s
806:	learn: 5533.3930689	total: 10.7s	remaining: 2.57s
807:	learn: 

951:	learn: 5443.9842967	total: 13.1s	remaining: 660ms
952:	learn: 5442.7688539	total: 13.1s	remaining: 646ms
953:	learn: 5442.5719691	total: 13.1s	remaining: 632ms
954:	learn: 5442.0572268	total: 13.1s	remaining: 618ms
955:	learn: 5441.5303177	total: 13.1s	remaining: 604ms
956:	learn: 5440.9435552	total: 13.1s	remaining: 590ms
957:	learn: 5440.5542146	total: 13.1s	remaining: 576ms
958:	learn: 5439.8329041	total: 13.1s	remaining: 562ms
959:	learn: 5439.4437634	total: 13.2s	remaining: 548ms
960:	learn: 5437.9892489	total: 13.2s	remaining: 534ms
961:	learn: 5437.5468629	total: 13.2s	remaining: 520ms
962:	learn: 5437.0718619	total: 13.2s	remaining: 506ms
963:	learn: 5436.8183233	total: 13.2s	remaining: 493ms
964:	learn: 5436.2902990	total: 13.2s	remaining: 479ms
965:	learn: 5435.9403191	total: 13.2s	remaining: 465ms
966:	learn: 5435.2942956	total: 13.2s	remaining: 451ms
967:	learn: 5434.9282447	total: 13.2s	remaining: 437ms
968:	learn: 5434.4926875	total: 13.2s	remaining: 423ms
969:	learn

`CAT Boost`
* r2_score =  0.7510773791747923

# Hyper-parameter Tuning

In [14]:
cbr = CatBoostRegressor(random_state=2021,iterations=250)
grid = {
    'learning_rate': [0.03, 0.1],
    'depth':[4, 6, 10],
    'l2_leaf_reg': [1, 3, 5]}
cbr.grid_search(grid,X_train,y_train, verbose=3, plot=True)
y_pred =cbr.predict(X_test)
print("r2_score = ",r2_score(y_test, y_pred))


MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 17900.3057863	test: 17984.6840070	best: 17984.6840070 (0)	total: 8.4ms	remaining: 2.09s
1:	learn: 17477.5548615	test: 17562.2470855	best: 17562.2470855 (1)	total: 12.7ms	remaining: 1.57s
2:	learn: 17080.3450506	test: 17163.0312108	best: 17163.0312108 (2)	total: 16.8ms	remaining: 1.39s
3:	learn: 16685.0992196	test: 16768.5082146	best: 16768.5082146 (3)	total: 21.2ms	remaining: 1.3s
4:	learn: 16314.7959164	test: 16397.7248979	best: 16397.7248979 (4)	total: 26.2ms	remaining: 1.28s
5:	learn: 15951.9805317	test: 16034.6644464	best: 16034.6644464 (5)	total: 30.7ms	remaining: 1.25s
6:	learn: 15607.3009407	test: 15690.8617279	best: 15690.8617279 (6)	total: 35.2ms	remaining: 1.22s
7:	learn: 15264.1308928	test: 15348.0135824	best: 15348.0135824 (7)	total: 39.6ms	remaining: 1.2s
8:	learn: 14930.9982418	test: 15014.8858314	best: 15014.8858314 (8)	total: 43.9ms	remaining: 1.18s
9:	learn: 14620.6410946	test: 14704.5244915	best: 14704.5244915 (9)	total: 48.2ms	remaining: 1.16s
10:	learn: 14

84:	learn: 7311.4492811	test: 7378.7861599	best: 7378.7861599 (84)	total: 775ms	remaining: 1.5s
85:	learn: 7296.4914479	test: 7363.7146229	best: 7363.7146229 (85)	total: 784ms	remaining: 1.49s
86:	learn: 7282.5305625	test: 7348.9208281	best: 7348.9208281 (86)	total: 794ms	remaining: 1.49s
87:	learn: 7266.7719775	test: 7332.2524446	best: 7332.2524446 (87)	total: 804ms	remaining: 1.48s
88:	learn: 7252.6916314	test: 7316.8862039	best: 7316.8862039 (88)	total: 813ms	remaining: 1.47s
89:	learn: 7238.6295063	test: 7304.3801122	best: 7304.3801122 (89)	total: 834ms	remaining: 1.48s
90:	learn: 7226.6810318	test: 7294.1851365	best: 7294.1851365 (90)	total: 843ms	remaining: 1.47s
91:	learn: 7212.6754392	test: 7280.7997376	best: 7280.7997376 (91)	total: 855ms	remaining: 1.47s
92:	learn: 7200.7863652	test: 7270.8706284	best: 7270.8706284 (92)	total: 873ms	remaining: 1.47s
93:	learn: 7188.2218945	test: 7258.7300169	best: 7258.7300169 (93)	total: 884ms	remaining: 1.47s
94:	learn: 7177.0850042	test: 7

178:	learn: 6669.1275718	test: 6732.5388926	best: 6732.5388926 (178)	total: 1.74s	remaining: 689ms
179:	learn: 6665.9708528	test: 6729.0363809	best: 6729.0363809 (179)	total: 1.75s	remaining: 679ms
180:	learn: 6663.3248373	test: 6726.1951550	best: 6726.1951550 (180)	total: 1.76s	remaining: 670ms
181:	learn: 6660.5949441	test: 6723.5445061	best: 6723.5445061 (181)	total: 1.77s	remaining: 660ms
182:	learn: 6657.1982806	test: 6719.5509251	best: 6719.5509251 (182)	total: 1.78s	remaining: 651ms
183:	learn: 6654.5381241	test: 6716.6688964	best: 6716.6688964 (183)	total: 1.79s	remaining: 641ms
184:	learn: 6649.6300115	test: 6711.7149589	best: 6711.7149589 (184)	total: 1.8s	remaining: 632ms
185:	learn: 6646.9786644	test: 6708.1614235	best: 6708.1614235 (185)	total: 1.8s	remaining: 621ms
186:	learn: 6644.5189846	test: 6705.5603324	best: 6705.5603324 (186)	total: 1.81s	remaining: 611ms
187:	learn: 6640.7511447	test: 6702.5267021	best: 6702.5267021 (187)	total: 1.82s	remaining: 601ms
188:	learn: 

20:	learn: 7566.6704695	test: 7635.2177551	best: 7635.2177551 (20)	total: 243ms	remaining: 2.65s
21:	learn: 7492.4052000	test: 7558.3062815	best: 7558.3062815 (21)	total: 254ms	remaining: 2.64s
22:	learn: 7429.1094474	test: 7488.7891454	best: 7488.7891454 (22)	total: 264ms	remaining: 2.6s
23:	learn: 7372.4044136	test: 7437.1620338	best: 7437.1620338 (23)	total: 278ms	remaining: 2.62s
24:	learn: 7315.8108659	test: 7378.5715605	best: 7378.5715605 (24)	total: 288ms	remaining: 2.59s
25:	learn: 7266.0959873	test: 7330.4781935	best: 7330.4781935 (25)	total: 298ms	remaining: 2.57s
26:	learn: 7220.4015109	test: 7282.2676486	best: 7282.2676486 (26)	total: 310ms	remaining: 2.56s
27:	learn: 7182.1158033	test: 7241.2393065	best: 7241.2393065 (27)	total: 319ms	remaining: 2.53s
28:	learn: 7143.1635383	test: 7200.1934179	best: 7200.1934179 (28)	total: 330ms	remaining: 2.51s
29:	learn: 7106.7999687	test: 7162.7579380	best: 7162.7579380 (29)	total: 339ms	remaining: 2.48s
30:	learn: 7069.6381847	test: 7

125:	learn: 6217.4222184	test: 6267.0498630	best: 6267.0498630 (125)	total: 1.19s	remaining: 1.17s
126:	learn: 6213.7247888	test: 6262.8949549	best: 6262.8949549 (126)	total: 1.19s	remaining: 1.16s
127:	learn: 6210.1278061	test: 6258.4173801	best: 6258.4173801 (127)	total: 1.2s	remaining: 1.14s
128:	learn: 6207.9441108	test: 6256.3045722	best: 6256.3045722 (128)	total: 1.2s	remaining: 1.13s
129:	learn: 6202.2130717	test: 6250.1436999	best: 6250.1436999 (129)	total: 1.21s	remaining: 1.11s
130:	learn: 6198.2772184	test: 6245.6444121	best: 6245.6444121 (130)	total: 1.21s	remaining: 1.1s
131:	learn: 6196.0555622	test: 6243.0617298	best: 6243.0617298 (131)	total: 1.22s	remaining: 1.09s
132:	learn: 6191.8512037	test: 6239.1022727	best: 6239.1022727 (132)	total: 1.22s	remaining: 1.07s
133:	learn: 6189.2839232	test: 6236.1829638	best: 6236.1829638 (133)	total: 1.23s	remaining: 1.06s
134:	learn: 6185.9602762	test: 6232.4043948	best: 6232.4043948 (134)	total: 1.23s	remaining: 1.05s
135:	learn: 6

233:	learn: 5914.1025315	test: 5984.8183336	best: 5984.8183336 (233)	total: 1.73s	remaining: 118ms
234:	learn: 5912.6030684	test: 5983.0645767	best: 5983.0645767 (234)	total: 1.74s	remaining: 111ms
235:	learn: 5910.4330767	test: 5981.4866917	best: 5981.4866917 (235)	total: 1.75s	remaining: 104ms
236:	learn: 5908.9350478	test: 5980.1196662	best: 5980.1196662 (236)	total: 1.75s	remaining: 96.1ms
237:	learn: 5907.5645682	test: 5978.7450450	best: 5978.7450450 (237)	total: 1.76s	remaining: 88.6ms
238:	learn: 5905.4633914	test: 5976.1353460	best: 5976.1353460 (238)	total: 1.76s	remaining: 81.1ms
239:	learn: 5901.1309419	test: 5971.7569603	best: 5971.7569603 (239)	total: 1.76s	remaining: 73.6ms
240:	learn: 5897.1044601	test: 5969.2576294	best: 5969.2576294 (240)	total: 1.77s	remaining: 66.1ms
241:	learn: 5895.8508282	test: 5969.0003304	best: 5969.0003304 (241)	total: 1.77s	remaining: 58.7ms
242:	learn: 5893.4640415	test: 5966.3655383	best: 5966.3655383 (242)	total: 1.78s	remaining: 51.2ms
243

69:	learn: 7620.6549232	test: 7670.4256934	best: 7670.4256934 (69)	total: 446ms	remaining: 1.15s
70:	learn: 7595.7415523	test: 7647.3640497	best: 7647.3640497 (70)	total: 457ms	remaining: 1.15s
71:	learn: 7569.6960761	test: 7620.8785354	best: 7620.8785354 (71)	total: 468ms	remaining: 1.16s
72:	learn: 7545.1344761	test: 7598.9474585	best: 7598.9474585 (72)	total: 478ms	remaining: 1.16s
73:	learn: 7523.8067309	test: 7576.9624625	best: 7576.9624625 (73)	total: 490ms	remaining: 1.16s
74:	learn: 7501.1677291	test: 7553.7303793	best: 7553.7303793 (74)	total: 500ms	remaining: 1.17s
75:	learn: 7480.3539186	test: 7534.6150972	best: 7534.6150972 (75)	total: 511ms	remaining: 1.17s
76:	learn: 7460.6634889	test: 7517.0499348	best: 7517.0499348 (76)	total: 521ms	remaining: 1.17s
77:	learn: 7440.2573743	test: 7498.9099108	best: 7498.9099108 (77)	total: 531ms	remaining: 1.17s
78:	learn: 7421.6168393	test: 7482.1789188	best: 7482.1789188 (78)	total: 542ms	remaining: 1.17s
79:	learn: 7404.6208011	test: 

165:	learn: 6733.9568524	test: 6798.0876252	best: 6798.0876252 (165)	total: 1.4s	remaining: 711ms
166:	learn: 6731.0620582	test: 6795.2781057	best: 6795.2781057 (166)	total: 1.42s	remaining: 704ms
167:	learn: 6726.0067504	test: 6790.0307930	best: 6790.0307930 (167)	total: 1.43s	remaining: 696ms
168:	learn: 6722.1937367	test: 6786.2328128	best: 6786.2328128 (168)	total: 1.44s	remaining: 688ms
169:	learn: 6718.4578850	test: 6782.9227936	best: 6782.9227936 (169)	total: 1.45s	remaining: 682ms
170:	learn: 6714.0531514	test: 6778.3021713	best: 6778.3021713 (170)	total: 1.46s	remaining: 673ms
171:	learn: 6710.1828528	test: 6774.3482109	best: 6774.3482109 (171)	total: 1.47s	remaining: 665ms
172:	learn: 6707.3006942	test: 6771.0638717	best: 6771.0638717 (172)	total: 1.48s	remaining: 657ms
173:	learn: 6704.3839786	test: 6767.3081332	best: 6767.3081332 (173)	total: 1.49s	remaining: 651ms
174:	learn: 6701.0364214	test: 6763.7663506	best: 6763.7663506 (174)	total: 1.5s	remaining: 643ms
175:	learn: 

13:	learn: 8532.4475294	test: 8592.0814642	best: 8592.0814642 (13)	total: 131ms	remaining: 2.21s
14:	learn: 8328.5709854	test: 8383.1705274	best: 8383.1705274 (14)	total: 141ms	remaining: 2.21s
15:	learn: 8150.1177299	test: 8204.1977779	best: 8204.1977779 (15)	total: 151ms	remaining: 2.21s
16:	learn: 7998.3106979	test: 8049.8116406	best: 8049.8116406 (16)	total: 160ms	remaining: 2.19s
17:	learn: 7863.4628144	test: 7911.8169688	best: 7911.8169688 (17)	total: 170ms	remaining: 2.19s
18:	learn: 7745.8159283	test: 7793.4205543	best: 7793.4205543 (18)	total: 179ms	remaining: 2.17s
19:	learn: 7648.6484318	test: 7695.7828933	best: 7695.7828933 (19)	total: 187ms	remaining: 2.15s
20:	learn: 7568.1725790	test: 7621.4834624	best: 7621.4834624 (20)	total: 198ms	remaining: 2.16s
21:	learn: 7492.4026677	test: 7546.4122464	best: 7546.4122464 (21)	total: 207ms	remaining: 2.14s
22:	learn: 7428.3322823	test: 7476.5245043	best: 7476.5245043 (22)	total: 217ms	remaining: 2.15s
23:	learn: 7371.2506292	test: 

114:	learn: 6280.2423783	test: 6321.4352939	best: 6321.4352939 (114)	total: 1.1s	remaining: 1.3s
115:	learn: 6275.6472516	test: 6317.0870803	best: 6317.0870803 (115)	total: 1.11s	remaining: 1.29s
116:	learn: 6271.9902924	test: 6313.4937712	best: 6313.4937712 (116)	total: 1.12s	remaining: 1.28s
117:	learn: 6269.7905126	test: 6310.5263844	best: 6310.5263844 (117)	total: 1.13s	remaining: 1.27s
118:	learn: 6264.9651525	test: 6304.0662073	best: 6304.0662073 (118)	total: 1.14s	remaining: 1.26s
119:	learn: 6262.2123941	test: 6300.9946878	best: 6300.9946878 (119)	total: 1.15s	remaining: 1.25s
120:	learn: 6255.2780696	test: 6292.3795647	best: 6292.3795647 (120)	total: 1.16s	remaining: 1.24s
121:	learn: 6251.4403235	test: 6288.0828669	best: 6288.0828669 (121)	total: 1.18s	remaining: 1.24s
122:	learn: 6247.0442116	test: 6283.4150710	best: 6283.4150710 (122)	total: 1.19s	remaining: 1.23s
123:	learn: 6241.7666599	test: 6279.6991879	best: 6279.6991879 (123)	total: 1.2s	remaining: 1.22s
124:	learn: 6

206:	learn: 5987.6142957	test: 6040.0279411	best: 6040.0279411 (206)	total: 2.04s	remaining: 424ms
207:	learn: 5985.4944484	test: 6037.6566845	best: 6037.6566845 (207)	total: 2.05s	remaining: 414ms
208:	learn: 5983.3810414	test: 6035.7443614	best: 6035.7443614 (208)	total: 2.06s	remaining: 404ms
209:	learn: 5982.3870223	test: 6034.7309703	best: 6034.7309703 (209)	total: 2.06s	remaining: 393ms
210:	learn: 5979.4427623	test: 6034.2224195	best: 6034.2224195 (210)	total: 2.07s	remaining: 383ms
211:	learn: 5978.3324866	test: 6033.7213170	best: 6033.7213170 (211)	total: 2.08s	remaining: 373ms
212:	learn: 5973.6673285	test: 6030.6749617	best: 6030.6749617 (212)	total: 2.09s	remaining: 364ms
213:	learn: 5972.9334858	test: 6029.7566072	best: 6029.7566072 (213)	total: 2.1s	remaining: 354ms
214:	learn: 5970.9646744	test: 6029.0793897	best: 6029.0793897 (214)	total: 2.11s	remaining: 344ms
215:	learn: 5968.9250043	test: 6026.9213617	best: 6026.9213617 (215)	total: 2.12s	remaining: 334ms
216:	learn:

56:	learn: 8068.6372479	test: 8121.4127110	best: 8121.4127110 (56)	total: 518ms	remaining: 1.75s
57:	learn: 8026.4886900	test: 8077.8500966	best: 8077.8500966 (57)	total: 527ms	remaining: 1.74s
58:	learn: 7985.2468906	test: 8035.8893065	best: 8035.8893065 (58)	total: 537ms	remaining: 1.74s
59:	learn: 7943.9860193	test: 7992.8632971	best: 7992.8632971 (59)	total: 547ms	remaining: 1.73s
60:	learn: 7905.3611247	test: 7954.6476703	best: 7954.6476703 (60)	total: 555ms	remaining: 1.72s
61:	learn: 7868.9608626	test: 7918.1554112	best: 7918.1554112 (61)	total: 564ms	remaining: 1.71s
62:	learn: 7833.4042093	test: 7881.5841708	best: 7881.5841708 (62)	total: 573ms	remaining: 1.7s
63:	learn: 7800.3561232	test: 7848.0914465	best: 7848.0914465 (63)	total: 583ms	remaining: 1.69s
64:	learn: 7767.0560310	test: 7814.6541132	best: 7814.6541132 (64)	total: 601ms	remaining: 1.71s
65:	learn: 7735.2913773	test: 7782.9183795	best: 7782.9183795 (65)	total: 610ms	remaining: 1.7s
66:	learn: 7706.1076732	test: 77

151:	learn: 6802.5101116	test: 6864.5447447	best: 6864.5447447 (151)	total: 1.46s	remaining: 941ms
152:	learn: 6797.8657872	test: 6860.2466320	best: 6860.2466320 (152)	total: 1.47s	remaining: 933ms
153:	learn: 6793.0484601	test: 6854.5448261	best: 6854.5448261 (153)	total: 1.48s	remaining: 923ms
154:	learn: 6789.2481954	test: 6850.3375968	best: 6850.3375968 (154)	total: 1.49s	remaining: 915ms
155:	learn: 6785.2801951	test: 6846.4746366	best: 6846.4746366 (155)	total: 1.5s	remaining: 905ms
156:	learn: 6781.7428560	test: 6843.2668140	best: 6843.2668140 (156)	total: 1.51s	remaining: 896ms
157:	learn: 6776.2325103	test: 6838.1550635	best: 6838.1550635 (157)	total: 1.52s	remaining: 888ms
158:	learn: 6771.7314009	test: 6834.3059017	best: 6834.3059017 (158)	total: 1.54s	remaining: 880ms
159:	learn: 6767.1836779	test: 6829.3096045	best: 6829.3096045 (159)	total: 1.55s	remaining: 870ms
160:	learn: 6764.0749907	test: 6825.4655299	best: 6825.4655299 (160)	total: 1.56s	remaining: 860ms
161:	learn:


bestTest = 6580.409122
bestIteration = 249

0:	learn: 16894.1490578	test: 16981.7323470	best: 16981.7323470 (0)	total: 11.9ms	remaining: 2.96s
1:	learn: 15618.3934033	test: 15706.1901479	best: 15706.1901479 (1)	total: 20.4ms	remaining: 2.54s
2:	learn: 14521.9983003	test: 14603.5342726	best: 14603.5342726 (2)	total: 29.5ms	remaining: 2.43s
3:	learn: 13532.1984840	test: 13614.6342443	best: 13614.6342443 (3)	total: 39.1ms	remaining: 2.4s
4:	learn: 12661.4237096	test: 12740.8929749	best: 12740.8929749 (4)	total: 48.2ms	remaining: 2.36s
5:	learn: 11899.8772182	test: 11978.2764076	best: 11978.2764076 (5)	total: 58.1ms	remaining: 2.36s
6:	learn: 11250.7205088	test: 11330.0740861	best: 11330.0740861 (6)	total: 66ms	remaining: 2.29s
7:	learn: 10668.7351616	test: 10743.7097274	best: 10743.7097274 (7)	total: 74.9ms	remaining: 2.27s
8:	learn: 10178.4236773	test: 10253.5536289	best: 10253.5536289 (8)	total: 83.4ms	remaining: 2.23s
9:	learn: 9741.8325550	test: 9812.3487010	best: 9812.3487010 (9)	to

95:	learn: 6391.9619930	test: 6446.0807152	best: 6446.0807152 (95)	total: 938ms	remaining: 1.5s
96:	learn: 6387.4025253	test: 6440.7180983	best: 6440.7180983 (96)	total: 950ms	remaining: 1.5s
97:	learn: 6384.2792748	test: 6436.5143484	best: 6436.5143484 (97)	total: 960ms	remaining: 1.49s
98:	learn: 6375.9978689	test: 6426.2580750	best: 6426.2580750 (98)	total: 970ms	remaining: 1.48s
99:	learn: 6370.2823144	test: 6420.1647021	best: 6420.1647021 (99)	total: 982ms	remaining: 1.47s
100:	learn: 6365.5841406	test: 6414.3399684	best: 6414.3399684 (100)	total: 994ms	remaining: 1.47s
101:	learn: 6357.5924384	test: 6408.3085067	best: 6408.3085067 (101)	total: 1s	remaining: 1.46s
102:	learn: 6354.1198570	test: 6404.3726852	best: 6404.3726852 (102)	total: 1.02s	remaining: 1.45s
103:	learn: 6348.9142801	test: 6402.0802716	best: 6402.0802716 (103)	total: 1.04s	remaining: 1.45s
104:	learn: 6343.0825421	test: 6396.4849885	best: 6396.4849885 (104)	total: 1.07s	remaining: 1.47s
105:	learn: 6337.2588506	

190:	learn: 6069.6439416	test: 6120.9971530	best: 6120.9971530 (190)	total: 1.88s	remaining: 582ms
191:	learn: 6067.7471040	test: 6119.4541360	best: 6119.4541360 (191)	total: 1.9s	remaining: 572ms
192:	learn: 6063.7598545	test: 6114.6978280	best: 6114.6978280 (192)	total: 1.9s	remaining: 562ms
193:	learn: 6061.2949726	test: 6112.0702630	best: 6112.0702630 (193)	total: 1.91s	remaining: 553ms
194:	learn: 6058.2074107	test: 6108.2750439	best: 6108.2750439 (194)	total: 1.93s	remaining: 543ms
195:	learn: 6056.3266089	test: 6106.3275946	best: 6106.3275946 (195)	total: 1.94s	remaining: 533ms
196:	learn: 6054.2342911	test: 6104.6300614	best: 6104.6300614 (196)	total: 1.95s	remaining: 523ms
197:	learn: 6048.9755630	test: 6100.2037103	best: 6100.2037103 (197)	total: 1.95s	remaining: 513ms
198:	learn: 6046.6465970	test: 6098.0270170	best: 6098.0270170 (198)	total: 1.96s	remaining: 503ms
199:	learn: 6044.8955396	test: 6095.9476244	best: 6095.9476244 (199)	total: 1.97s	remaining: 493ms
200:	learn: 

29:	learn: 10091.4970844	test: 10163.1321771	best: 10163.1321771 (29)	total: 394ms	remaining: 2.89s
30:	learn: 9952.7468822	test: 10023.5361734	best: 10023.5361734 (30)	total: 408ms	remaining: 2.88s
31:	learn: 9816.3981412	test: 9886.7574096	best: 9886.7574096 (31)	total: 422ms	remaining: 2.87s
32:	learn: 9684.4809191	test: 9754.1286188	best: 9754.1286188 (32)	total: 434ms	remaining: 2.85s
33:	learn: 9559.8056752	test: 9628.6300499	best: 9628.6300499 (33)	total: 449ms	remaining: 2.85s
34:	learn: 9439.4144850	test: 9507.7129705	best: 9507.7129705 (34)	total: 463ms	remaining: 2.85s
35:	learn: 9327.8905365	test: 9396.3085873	best: 9396.3085873 (35)	total: 481ms	remaining: 2.86s
36:	learn: 9215.1245012	test: 9282.7055402	best: 9282.7055402 (36)	total: 497ms	remaining: 2.86s
37:	learn: 9107.0004465	test: 9175.1155835	best: 9175.1155835 (37)	total: 510ms	remaining: 2.85s
38:	learn: 9004.6645714	test: 9071.3028042	best: 9071.3028042 (38)	total: 523ms	remaining: 2.83s
39:	learn: 8907.1229289	t

117:	learn: 6645.2435719	test: 6720.3649456	best: 6720.3649456 (117)	total: 1.55s	remaining: 1.74s
118:	learn: 6638.2859567	test: 6713.2908939	best: 6713.2908939 (118)	total: 1.57s	remaining: 1.73s
119:	learn: 6630.9040339	test: 6707.8443786	best: 6707.8443786 (119)	total: 1.58s	remaining: 1.71s
120:	learn: 6621.8476456	test: 6698.6002748	best: 6698.6002748 (120)	total: 1.59s	remaining: 1.7s
121:	learn: 6612.1364120	test: 6688.1754299	best: 6688.1754299 (121)	total: 1.6s	remaining: 1.68s
122:	learn: 6605.0201627	test: 6681.1738445	best: 6681.1738445 (122)	total: 1.61s	remaining: 1.67s
123:	learn: 6598.2180251	test: 6673.7272693	best: 6673.7272693 (123)	total: 1.63s	remaining: 1.65s
124:	learn: 6592.0897177	test: 6667.1496462	best: 6667.1496462 (124)	total: 1.64s	remaining: 1.64s
125:	learn: 6586.6796863	test: 6661.2897827	best: 6661.2897827 (125)	total: 1.65s	remaining: 1.63s
126:	learn: 6581.1723789	test: 6655.1078591	best: 6655.1078591 (126)	total: 1.66s	remaining: 1.61s
127:	learn: 

209:	learn: 6266.2041722	test: 6335.0975891	best: 6335.0975891 (209)	total: 2.74s	remaining: 521ms
210:	learn: 6260.9584052	test: 6330.9134078	best: 6330.9134078 (210)	total: 2.75s	remaining: 508ms
211:	learn: 6258.8536632	test: 6328.6180354	best: 6328.6180354 (211)	total: 2.76s	remaining: 496ms
212:	learn: 6257.0976206	test: 6326.9669858	best: 6326.9669858 (212)	total: 2.78s	remaining: 482ms
213:	learn: 6255.1519302	test: 6324.8167516	best: 6324.8167516 (213)	total: 2.79s	remaining: 469ms
214:	learn: 6252.1565413	test: 6321.4860241	best: 6321.4860241 (214)	total: 2.8s	remaining: 456ms
215:	learn: 6249.0708765	test: 6318.7241501	best: 6318.7241501 (215)	total: 2.81s	remaining: 443ms
216:	learn: 6245.9809886	test: 6316.0200778	best: 6316.0200778 (216)	total: 2.82s	remaining: 429ms
217:	learn: 6244.1260079	test: 6314.1763400	best: 6314.1763400 (217)	total: 2.84s	remaining: 416ms
218:	learn: 6242.6979361	test: 6312.5672742	best: 6312.5672742 (218)	total: 2.85s	remaining: 403ms
219:	learn:

43:	learn: 6478.0320905	test: 6563.2547249	best: 6563.2547249 (43)	total: 625ms	remaining: 2.92s
44:	learn: 6465.8677159	test: 6551.7493611	best: 6551.7493611 (44)	total: 642ms	remaining: 2.92s
45:	learn: 6449.0739455	test: 6534.0111203	best: 6534.0111203 (45)	total: 656ms	remaining: 2.91s
46:	learn: 6434.3959192	test: 6519.4711252	best: 6519.4711252 (46)	total: 668ms	remaining: 2.88s
47:	learn: 6419.8130853	test: 6502.5543626	best: 6502.5543626 (47)	total: 685ms	remaining: 2.88s
48:	learn: 6406.8412807	test: 6490.6798942	best: 6490.6798942 (48)	total: 699ms	remaining: 2.87s
49:	learn: 6394.5043281	test: 6476.7844304	best: 6476.7844304 (49)	total: 709ms	remaining: 2.83s
50:	learn: 6384.2676931	test: 6466.9893299	best: 6466.9893299 (50)	total: 729ms	remaining: 2.85s
51:	learn: 6375.0880801	test: 6455.6814048	best: 6455.6814048 (51)	total: 742ms	remaining: 2.83s
52:	learn: 6362.6231825	test: 6441.7868462	best: 6441.7868462 (52)	total: 759ms	remaining: 2.82s
53:	learn: 6353.4158688	test: 

134:	learn: 5842.1572669	test: 5959.0806457	best: 5959.0806457 (134)	total: 2.02s	remaining: 1.72s
135:	learn: 5836.4665961	test: 5950.4779354	best: 5950.4779354 (135)	total: 2.04s	remaining: 1.71s
136:	learn: 5830.7641587	test: 5944.6872320	best: 5944.6872320 (136)	total: 2.05s	remaining: 1.69s
137:	learn: 5825.3960713	test: 5939.1937893	best: 5939.1937893 (137)	total: 2.06s	remaining: 1.68s
138:	learn: 5820.4521607	test: 5935.9625307	best: 5935.9625307 (138)	total: 2.08s	remaining: 1.66s
139:	learn: 5815.4483455	test: 5935.2537842	best: 5935.2537842 (139)	total: 2.09s	remaining: 1.64s
140:	learn: 5812.6632862	test: 5933.1743665	best: 5933.1743665 (140)	total: 2.1s	remaining: 1.63s
141:	learn: 5809.0310738	test: 5930.4549726	best: 5930.4549726 (141)	total: 2.12s	remaining: 1.61s
142:	learn: 5804.2105592	test: 5926.6730988	best: 5926.6730988 (142)	total: 2.13s	remaining: 1.59s
143:	learn: 5800.2089724	test: 5922.5178729	best: 5922.5178729 (143)	total: 2.15s	remaining: 1.58s
144:	learn:

222:	learn: 5517.0700159	test: 5704.6075574	best: 5704.6075574 (222)	total: 3.2s	remaining: 388ms
223:	learn: 5512.6683117	test: 5699.3517052	best: 5699.3517052 (223)	total: 3.22s	remaining: 373ms
224:	learn: 5510.2942680	test: 5697.6581649	best: 5697.6581649 (224)	total: 3.23s	remaining: 359ms
225:	learn: 5508.9710126	test: 5697.6645898	best: 5697.6581649 (224)	total: 3.24s	remaining: 344ms
226:	learn: 5507.1114540	test: 5696.2184443	best: 5696.2184443 (226)	total: 3.25s	remaining: 330ms
227:	learn: 5504.3156990	test: 5693.9691933	best: 5693.9691933 (227)	total: 3.27s	remaining: 315ms
228:	learn: 5501.4821423	test: 5693.1209642	best: 5693.1209642 (228)	total: 3.28s	remaining: 300ms
229:	learn: 5499.2331897	test: 5691.6198766	best: 5691.6198766 (229)	total: 3.29s	remaining: 286ms
230:	learn: 5497.2345397	test: 5689.5268941	best: 5689.5268941 (230)	total: 3.3s	remaining: 271ms
231:	learn: 5493.8401693	test: 5687.3350660	best: 5687.3350660 (231)	total: 3.31s	remaining: 257ms
232:	learn: 

62:	learn: 7534.9772883	test: 7601.1937656	best: 7601.1937656 (62)	total: 811ms	remaining: 2.41s
63:	learn: 7500.8131622	test: 7568.0880050	best: 7568.0880050 (63)	total: 824ms	remaining: 2.39s
64:	learn: 7467.1996603	test: 7535.2939710	best: 7535.2939710 (64)	total: 837ms	remaining: 2.38s
65:	learn: 7436.1512743	test: 7506.1807973	best: 7506.1807973 (65)	total: 853ms	remaining: 2.38s
66:	learn: 7406.1697748	test: 7478.0944195	best: 7478.0944195 (66)	total: 866ms	remaining: 2.36s
67:	learn: 7376.4067541	test: 7448.6619920	best: 7448.6619920 (67)	total: 878ms	remaining: 2.35s
68:	learn: 7346.0994015	test: 7419.7334103	best: 7419.7334103 (68)	total: 898ms	remaining: 2.36s
69:	learn: 7315.6620308	test: 7389.7916892	best: 7389.7916892 (69)	total: 910ms	remaining: 2.34s
70:	learn: 7289.9328335	test: 7365.5217390	best: 7365.5217390 (70)	total: 922ms	remaining: 2.32s
71:	learn: 7263.2295915	test: 7338.3383271	best: 7338.3383271 (71)	total: 936ms	remaining: 2.31s
72:	learn: 7239.2614463	test: 

151:	learn: 6491.3994617	test: 6558.6392521	best: 6558.6392521 (151)	total: 2s	remaining: 1.29s
152:	learn: 6486.0041701	test: 6553.5839424	best: 6553.5839424 (152)	total: 2.02s	remaining: 1.28s
153:	learn: 6480.7840712	test: 6548.1644591	best: 6548.1644591 (153)	total: 2.03s	remaining: 1.27s
154:	learn: 6475.7590821	test: 6543.2712269	best: 6543.2712269 (154)	total: 2.04s	remaining: 1.25s
155:	learn: 6472.6251077	test: 6540.4087819	best: 6540.4087819 (155)	total: 2.05s	remaining: 1.24s
156:	learn: 6468.4944488	test: 6535.4245973	best: 6535.4245973 (156)	total: 2.06s	remaining: 1.22s
157:	learn: 6464.9480069	test: 6532.6232030	best: 6532.6232030 (157)	total: 2.08s	remaining: 1.21s
158:	learn: 6461.2860583	test: 6528.7676950	best: 6528.7676950 (158)	total: 2.09s	remaining: 1.2s
159:	learn: 6455.9254768	test: 6522.3237221	best: 6522.3237221 (159)	total: 2.1s	remaining: 1.18s
160:	learn: 6451.3893850	test: 6517.6410038	best: 6517.6410038 (160)	total: 2.11s	remaining: 1.17s
161:	learn: 644

246:	learn: 6206.7908269	test: 6278.7433529	best: 6278.7433529 (246)	total: 3.18s	remaining: 38.6ms
247:	learn: 6204.8833681	test: 6276.9085898	best: 6276.9085898 (247)	total: 3.19s	remaining: 25.7ms
248:	learn: 6203.2051433	test: 6275.4334785	best: 6275.4334785 (248)	total: 3.2s	remaining: 12.9ms
249:	learn: 6201.4731537	test: 6273.3176619	best: 6273.3176619 (249)	total: 3.21s	remaining: 0us

bestTest = 6273.317662
bestIteration = 249

0:	learn: 16874.7104334	test: 16959.6985427	best: 16959.6985427 (0)	total: 14.3ms	remaining: 3.56s
1:	learn: 15559.0634483	test: 15636.3538721	best: 15636.3538721 (1)	total: 26.1ms	remaining: 3.24s
2:	learn: 14399.0697867	test: 14481.1004744	best: 14481.1004744 (2)	total: 38.7ms	remaining: 3.19s
3:	learn: 13394.6629505	test: 13474.4397217	best: 13474.4397217 (3)	total: 50.5ms	remaining: 3.1s
4:	learn: 12503.7317794	test: 12585.6063393	best: 12585.6063393 (4)	total: 65.4ms	remaining: 3.2s
5:	learn: 11743.0667774	test: 11824.1027321	best: 11824.1027321 (5

85:	learn: 6130.1388218	test: 6209.1788730	best: 6209.1788730 (85)	total: 1.13s	remaining: 2.15s
86:	learn: 6122.9164323	test: 6201.3018183	best: 6201.3018183 (86)	total: 1.14s	remaining: 2.14s
87:	learn: 6110.7234305	test: 6192.7154756	best: 6192.7154756 (87)	total: 1.16s	remaining: 2.13s
88:	learn: 6106.2838395	test: 6188.2040114	best: 6188.2040114 (88)	total: 1.17s	remaining: 2.12s
89:	learn: 6100.3369338	test: 6182.8428155	best: 6182.8428155 (89)	total: 1.18s	remaining: 2.1s
90:	learn: 6096.4630912	test: 6179.8600427	best: 6179.8600427 (90)	total: 1.2s	remaining: 2.09s
91:	learn: 6092.6122730	test: 6176.1174449	best: 6176.1174449 (91)	total: 1.21s	remaining: 2.07s
92:	learn: 6085.6984869	test: 6168.7312960	best: 6168.7312960 (92)	total: 1.22s	remaining: 2.06s
93:	learn: 6078.2199487	test: 6161.9977024	best: 6161.9977024 (93)	total: 1.23s	remaining: 2.04s
94:	learn: 6074.5309301	test: 6157.9790785	best: 6157.9790785 (94)	total: 1.24s	remaining: 2.03s
95:	learn: 6069.9548246	test: 61

182:	learn: 5692.7354936	test: 5826.4322244	best: 5826.4322244 (182)	total: 2.3s	remaining: 844ms
183:	learn: 5690.1853590	test: 5825.1385247	best: 5825.1385247 (183)	total: 2.32s	remaining: 831ms
184:	learn: 5686.0553105	test: 5823.9721693	best: 5823.9721693 (184)	total: 2.33s	remaining: 819ms
185:	learn: 5682.8338440	test: 5821.8422047	best: 5821.8422047 (185)	total: 2.34s	remaining: 806ms
186:	learn: 5680.2892817	test: 5819.6469456	best: 5819.6469456 (186)	total: 2.35s	remaining: 793ms
187:	learn: 5676.3977127	test: 5817.8038968	best: 5817.8038968 (187)	total: 2.37s	remaining: 780ms
188:	learn: 5673.1246991	test: 5813.5686074	best: 5813.5686074 (188)	total: 2.38s	remaining: 767ms
189:	learn: 5669.8109315	test: 5808.9320918	best: 5808.9320918 (189)	total: 2.39s	remaining: 754ms
190:	learn: 5668.3660997	test: 5808.2062681	best: 5808.2062681 (190)	total: 2.4s	remaining: 741ms
191:	learn: 5666.0481534	test: 5807.0209321	best: 5807.0209321 (191)	total: 2.41s	remaining: 728ms
192:	learn: 

21:	learn: 11475.2312526	test: 11552.5186278	best: 11552.5186278 (21)	total: 312ms	remaining: 3.23s
22:	learn: 11279.1851913	test: 11355.5476070	best: 11355.5476070 (22)	total: 325ms	remaining: 3.21s
23:	learn: 11091.7543893	test: 11167.3067016	best: 11167.3067016 (23)	total: 338ms	remaining: 3.18s
24:	learn: 10909.7112994	test: 10984.5283488	best: 10984.5283488 (24)	total: 351ms	remaining: 3.16s
25:	learn: 10736.5042333	test: 10810.4325077	best: 10810.4325077 (25)	total: 362ms	remaining: 3.12s
26:	learn: 10569.7404594	test: 10643.1435088	best: 10643.1435088 (26)	total: 374ms	remaining: 3.09s
27:	learn: 10409.9443145	test: 10481.7759683	best: 10481.7759683 (27)	total: 386ms	remaining: 3.06s
28:	learn: 10254.6520628	test: 10325.7220115	best: 10325.7220115 (28)	total: 398ms	remaining: 3.03s
29:	learn: 10111.3488980	test: 10182.1420604	best: 10182.1420604 (29)	total: 409ms	remaining: 3s
30:	learn: 9972.4260589	test: 10042.2075634	best: 10042.2075634 (30)	total: 420ms	remaining: 2.97s
31:	

114:	learn: 6711.6470216	test: 6777.6724111	best: 6777.6724111 (114)	total: 1.48s	remaining: 1.74s
115:	learn: 6705.0242327	test: 6770.5205395	best: 6770.5205395 (115)	total: 1.5s	remaining: 1.73s
116:	learn: 6699.4678228	test: 6764.5851805	best: 6764.5851805 (116)	total: 1.51s	remaining: 1.72s
117:	learn: 6693.0997598	test: 6757.8795135	best: 6757.8795135 (117)	total: 1.52s	remaining: 1.7s
118:	learn: 6687.5846318	test: 6751.7753011	best: 6751.7753011 (118)	total: 1.53s	remaining: 1.69s
119:	learn: 6681.3599718	test: 6746.2347997	best: 6746.2347997 (119)	total: 1.55s	remaining: 1.68s
120:	learn: 6675.9671511	test: 6739.9027918	best: 6739.9027918 (120)	total: 1.56s	remaining: 1.67s
121:	learn: 6668.5494061	test: 6733.0112143	best: 6733.0112143 (121)	total: 1.58s	remaining: 1.66s
122:	learn: 6662.7420998	test: 6727.4055942	best: 6727.4055942 (122)	total: 1.6s	remaining: 1.65s
123:	learn: 6656.6114632	test: 6720.9151555	best: 6720.9151555 (123)	total: 1.61s	remaining: 1.64s
124:	learn: 6

197:	learn: 6362.4355773	test: 6428.3409270	best: 6428.3409270 (197)	total: 2.66s	remaining: 699ms
198:	learn: 6360.2660046	test: 6426.1702511	best: 6426.1702511 (198)	total: 2.67s	remaining: 685ms
199:	learn: 6358.1700227	test: 6423.8405496	best: 6423.8405496 (199)	total: 2.68s	remaining: 671ms
200:	learn: 6355.1803027	test: 6420.5538245	best: 6420.5538245 (200)	total: 2.7s	remaining: 657ms
201:	learn: 6352.9316204	test: 6418.7226630	best: 6418.7226630 (201)	total: 2.71s	remaining: 644ms
202:	learn: 6350.0298834	test: 6415.6395648	best: 6415.6395648 (202)	total: 2.72s	remaining: 631ms
203:	learn: 6347.0781873	test: 6413.1804234	best: 6413.1804234 (203)	total: 2.74s	remaining: 617ms
204:	learn: 6343.8518564	test: 6409.9869430	best: 6409.9869430 (204)	total: 2.75s	remaining: 603ms
205:	learn: 6341.4655920	test: 6407.7854875	best: 6407.7854875 (205)	total: 2.76s	remaining: 589ms
206:	learn: 6339.2510038	test: 6404.9382654	best: 6404.9382654 (206)	total: 2.77s	remaining: 575ms
207:	learn:

35:	learn: 6669.5993783	test: 6736.7462097	best: 6736.7462097 (35)	total: 479ms	remaining: 2.85s
36:	learn: 6640.5612010	test: 6709.3583031	best: 6709.3583031 (36)	total: 493ms	remaining: 2.83s
37:	learn: 6624.7723901	test: 6696.5481239	best: 6696.5481239 (37)	total: 507ms	remaining: 2.83s
38:	learn: 6607.0653291	test: 6681.1959009	best: 6681.1959009 (38)	total: 521ms	remaining: 2.82s
39:	learn: 6590.1601997	test: 6661.7098820	best: 6661.7098820 (39)	total: 536ms	remaining: 2.82s
40:	learn: 6578.0647744	test: 6647.0666662	best: 6647.0666662 (40)	total: 552ms	remaining: 2.81s
41:	learn: 6564.4650848	test: 6631.5872254	best: 6631.5872254 (41)	total: 565ms	remaining: 2.79s
42:	learn: 6548.1314628	test: 6617.1535777	best: 6617.1535777 (42)	total: 577ms	remaining: 2.78s
43:	learn: 6532.3663592	test: 6605.2287380	best: 6605.2287380 (43)	total: 594ms	remaining: 2.78s
44:	learn: 6517.2842752	test: 6590.2616393	best: 6590.2616393 (44)	total: 608ms	remaining: 2.77s
45:	learn: 6501.6605667	test: 

134:	learn: 5920.4904094	test: 6004.5815269	best: 6004.5815269 (134)	total: 1.86s	remaining: 1.58s
135:	learn: 5916.8904849	test: 6001.6684621	best: 6001.6684621 (135)	total: 1.87s	remaining: 1.57s
136:	learn: 5911.9969107	test: 5998.2616612	best: 5998.2616612 (136)	total: 1.88s	remaining: 1.55s
137:	learn: 5908.9274353	test: 5996.2035964	best: 5996.2035964 (137)	total: 1.9s	remaining: 1.54s
138:	learn: 5906.2574688	test: 5993.9904784	best: 5993.9904784 (138)	total: 1.91s	remaining: 1.53s
139:	learn: 5903.6001670	test: 5991.9701417	best: 5991.9701417 (139)	total: 1.93s	remaining: 1.51s
140:	learn: 5900.1685496	test: 5989.5739515	best: 5989.5739515 (140)	total: 1.94s	remaining: 1.5s
141:	learn: 5894.8511254	test: 5985.4635543	best: 5985.4635543 (141)	total: 1.95s	remaining: 1.48s
142:	learn: 5888.6479029	test: 5980.0466333	best: 5980.0466333 (142)	total: 1.97s	remaining: 1.47s
143:	learn: 5885.3876202	test: 5977.0537912	best: 5977.0537912 (143)	total: 1.98s	remaining: 1.46s
144:	learn: 

225:	learn: 5632.3604163	test: 5765.0403570	best: 5765.0403570 (225)	total: 3s	remaining: 319ms
226:	learn: 5629.9480098	test: 5763.3444303	best: 5763.3444303 (226)	total: 3.02s	remaining: 306ms
227:	learn: 5628.2127859	test: 5761.8441555	best: 5761.8441555 (227)	total: 3.03s	remaining: 292ms
228:	learn: 5626.3192860	test: 5760.6383349	best: 5760.6383349 (228)	total: 3.04s	remaining: 279ms
229:	learn: 5621.5751653	test: 5758.5716968	best: 5758.5716968 (229)	total: 3.05s	remaining: 266ms
230:	learn: 5616.6405829	test: 5753.3447973	best: 5753.3447973 (230)	total: 3.07s	remaining: 252ms
231:	learn: 5613.7171725	test: 5749.8526807	best: 5749.8526807 (231)	total: 3.08s	remaining: 239ms
232:	learn: 5610.6330766	test: 5747.6617049	best: 5747.6617049 (232)	total: 3.09s	remaining: 225ms
233:	learn: 5607.7897774	test: 5743.8620534	best: 5743.8620534 (233)	total: 3.1s	remaining: 212ms
234:	learn: 5605.0661772	test: 5742.7879123	best: 5742.7879123 (234)	total: 3.11s	remaining: 199ms
235:	learn: 56

59:	learn: 7192.2059409	test: 7343.8174012	best: 7343.8174012 (59)	total: 2.64s	remaining: 8.37s
60:	learn: 7150.3425957	test: 7303.6569686	best: 7303.6569686 (60)	total: 2.68s	remaining: 8.31s
61:	learn: 7112.0678688	test: 7267.2961781	best: 7267.2961781 (61)	total: 2.72s	remaining: 8.25s
62:	learn: 7076.0726462	test: 7232.3848424	best: 7232.3848424 (62)	total: 2.76s	remaining: 8.19s
63:	learn: 7036.7228167	test: 7192.8085596	best: 7192.8085596 (63)	total: 2.8s	remaining: 8.14s
64:	learn: 7003.4259502	test: 7160.5659616	best: 7160.5659616 (64)	total: 2.84s	remaining: 8.09s
65:	learn: 6967.9951683	test: 7125.3797691	best: 7125.3797691 (65)	total: 2.89s	remaining: 8.06s
66:	learn: 6935.5230853	test: 7092.5042268	best: 7092.5042268 (66)	total: 2.94s	remaining: 8.02s
67:	learn: 6900.3883677	test: 7059.0160911	best: 7059.0160911 (67)	total: 2.98s	remaining: 7.98s
68:	learn: 6867.6948526	test: 7025.9292166	best: 7025.9292166 (68)	total: 3.03s	remaining: 7.95s
69:	learn: 6839.2935061	test: 6

147:	learn: 5938.5336545	test: 6126.9012871	best: 6126.9012871 (147)	total: 6.54s	remaining: 4.5s
148:	learn: 5934.3822145	test: 6122.6553130	best: 6122.6553130 (148)	total: 6.59s	remaining: 4.46s
149:	learn: 5928.5161709	test: 6116.0774841	best: 6116.0774841 (149)	total: 6.64s	remaining: 4.43s
150:	learn: 5924.6767204	test: 6110.2524379	best: 6110.2524379 (150)	total: 6.68s	remaining: 4.38s
151:	learn: 5920.5567480	test: 6106.0196608	best: 6106.0196608 (151)	total: 6.73s	remaining: 4.34s
152:	learn: 5916.6988266	test: 6101.4977500	best: 6101.4977500 (152)	total: 6.78s	remaining: 4.3s
153:	learn: 5913.6024755	test: 6098.2519301	best: 6098.2519301 (153)	total: 6.82s	remaining: 4.25s
154:	learn: 5910.2729122	test: 6094.6407991	best: 6094.6407991 (154)	total: 6.86s	remaining: 4.21s
155:	learn: 5902.9143454	test: 6090.5211173	best: 6090.5211173 (155)	total: 6.9s	remaining: 4.16s
156:	learn: 5898.9389125	test: 6085.8392828	best: 6085.8392828 (156)	total: 6.94s	remaining: 4.11s
157:	learn: 5

233:	learn: 5611.6223024	test: 5840.6785994	best: 5840.6785994 (233)	total: 10.4s	remaining: 714ms
234:	learn: 5609.0997707	test: 5838.0583824	best: 5838.0583824 (234)	total: 10.5s	remaining: 669ms
235:	learn: 5605.0460815	test: 5835.3598039	best: 5835.3598039 (235)	total: 10.5s	remaining: 624ms
236:	learn: 5602.1221046	test: 5832.7850840	best: 5832.7850840 (236)	total: 10.6s	remaining: 580ms
237:	learn: 5599.0204792	test: 5830.3159298	best: 5830.3159298 (237)	total: 10.6s	remaining: 535ms
238:	learn: 5594.3617380	test: 5827.7381774	best: 5827.7381774 (238)	total: 10.7s	remaining: 491ms
239:	learn: 5589.0073891	test: 5824.3790218	best: 5824.3790218 (239)	total: 10.7s	remaining: 446ms
240:	learn: 5586.9147372	test: 5822.3408874	best: 5822.3408874 (240)	total: 10.7s	remaining: 401ms
241:	learn: 5583.9284679	test: 5819.2124231	best: 5819.2124231 (241)	total: 10.8s	remaining: 356ms
242:	learn: 5579.4813270	test: 5815.8414122	best: 5815.8414122 (242)	total: 10.8s	remaining: 312ms
243:	learn

72:	learn: 5592.8835276	test: 5822.1319915	best: 5822.1319915 (72)	total: 3.29s	remaining: 7.97s
73:	learn: 5587.6591587	test: 5816.3096109	best: 5816.3096109 (73)	total: 3.34s	remaining: 7.93s
74:	learn: 5573.8991359	test: 5805.6558904	best: 5805.6558904 (74)	total: 3.38s	remaining: 7.9s
75:	learn: 5566.5376348	test: 5799.6181112	best: 5799.6181112 (75)	total: 3.43s	remaining: 7.86s
76:	learn: 5548.5390945	test: 5789.3290690	best: 5789.3290690 (76)	total: 3.48s	remaining: 7.82s
77:	learn: 5536.9632877	test: 5784.6984971	best: 5784.6984971 (77)	total: 3.53s	remaining: 7.78s
78:	learn: 5528.7327354	test: 5781.6663814	best: 5781.6663814 (78)	total: 3.57s	remaining: 7.72s
79:	learn: 5522.0041805	test: 5775.1614419	best: 5775.1614419 (79)	total: 3.61s	remaining: 7.67s
80:	learn: 5511.3108535	test: 5767.8011306	best: 5767.8011306 (80)	total: 3.65s	remaining: 7.61s
81:	learn: 5500.5855922	test: 5759.6088177	best: 5759.6088177 (81)	total: 3.69s	remaining: 7.56s
82:	learn: 5487.5435782	test: 5

157:	learn: 4931.6385499	test: 5410.3580560	best: 5410.3580560 (157)	total: 7.07s	remaining: 4.12s
158:	learn: 4925.7888964	test: 5407.5268559	best: 5407.5268559 (158)	total: 7.12s	remaining: 4.07s
159:	learn: 4923.0188515	test: 5405.8651122	best: 5405.8651122 (159)	total: 7.16s	remaining: 4.03s
160:	learn: 4917.2220118	test: 5401.0577873	best: 5401.0577873 (160)	total: 7.21s	remaining: 3.99s
161:	learn: 4912.2880588	test: 5397.6514665	best: 5397.6514665 (161)	total: 7.26s	remaining: 3.94s
162:	learn: 4908.9943900	test: 5396.1151712	best: 5396.1151712 (162)	total: 7.3s	remaining: 3.9s
163:	learn: 4904.4933874	test: 5394.9148040	best: 5394.9148040 (163)	total: 7.35s	remaining: 3.85s
164:	learn: 4900.0060849	test: 5390.0917262	best: 5390.0917262 (164)	total: 7.39s	remaining: 3.81s
165:	learn: 4891.4808267	test: 5385.2157866	best: 5385.2157866 (165)	total: 7.43s	remaining: 3.76s
166:	learn: 4887.7094794	test: 5382.2398906	best: 5382.2398906 (166)	total: 7.47s	remaining: 3.71s
167:	learn: 

241:	learn: 4534.0403154	test: 5221.7284981	best: 5221.7284981 (241)	total: 10.9s	remaining: 360ms
242:	learn: 4528.7648209	test: 5219.9685990	best: 5219.9685990 (242)	total: 10.9s	remaining: 315ms
243:	learn: 4525.7364301	test: 5217.7623988	best: 5217.7623988 (243)	total: 11s	remaining: 270ms
244:	learn: 4519.9837197	test: 5214.2247683	best: 5214.2247683 (244)	total: 11s	remaining: 225ms
245:	learn: 4515.7919921	test: 5213.4942984	best: 5213.4942984 (245)	total: 11.1s	remaining: 180ms
246:	learn: 4510.9466188	test: 5211.8218893	best: 5211.8218893 (246)	total: 11.1s	remaining: 135ms
247:	learn: 4506.8297075	test: 5210.3822587	best: 5210.3822587 (247)	total: 11.2s	remaining: 90ms
248:	learn: 4500.3026217	test: 5204.1173614	best: 5204.1173614 (248)	total: 11.2s	remaining: 45ms
249:	learn: 4497.4038163	test: 5202.4764285	best: 5202.4764285 (249)	total: 11.2s	remaining: 0us

bestTest = 5202.476429
bestIteration = 249

0:	learn: 17895.3476833	test: 17978.7953206	best: 17978.7953206 (0)	tota

83:	learn: 6585.2375829	test: 6732.3664256	best: 6732.3664256 (83)	total: 2.71s	remaining: 5.36s
84:	learn: 6568.3872104	test: 6715.7605490	best: 6715.7605490 (84)	total: 2.73s	remaining: 5.31s
85:	learn: 6553.1634632	test: 6701.7254726	best: 6701.7254726 (85)	total: 2.76s	remaining: 5.25s
86:	learn: 6537.5931354	test: 6685.0947329	best: 6685.0947329 (86)	total: 2.78s	remaining: 5.21s
87:	learn: 6520.6919911	test: 6667.2285992	best: 6667.2285992 (87)	total: 2.83s	remaining: 5.21s
88:	learn: 6504.6352740	test: 6652.3763257	best: 6652.3763257 (88)	total: 2.88s	remaining: 5.2s
89:	learn: 6489.2820799	test: 6638.4428792	best: 6638.4428792 (89)	total: 2.93s	remaining: 5.2s
90:	learn: 6472.8074434	test: 6622.6997479	best: 6622.6997479 (90)	total: 2.97s	remaining: 5.19s
91:	learn: 6459.5813462	test: 6610.3309918	best: 6610.3309918 (91)	total: 3.01s	remaining: 5.17s
92:	learn: 6446.7720888	test: 6596.2671057	best: 6596.2671057 (92)	total: 3.05s	remaining: 5.16s
93:	learn: 6435.2273843	test: 65

171:	learn: 5937.1940089	test: 6101.7232046	best: 6101.7232046 (171)	total: 6.88s	remaining: 3.12s
172:	learn: 5933.1040427	test: 6096.9998043	best: 6096.9998043 (172)	total: 6.92s	remaining: 3.08s
173:	learn: 5929.8699846	test: 6092.8541176	best: 6092.8541176 (173)	total: 6.97s	remaining: 3.04s
174:	learn: 5924.3357973	test: 6088.4919671	best: 6088.4919671 (174)	total: 7.02s	remaining: 3.01s
175:	learn: 5919.1702858	test: 6084.5582221	best: 6084.5582221 (175)	total: 7.08s	remaining: 2.98s
176:	learn: 5914.8528745	test: 6081.1432742	best: 6081.1432742 (176)	total: 7.14s	remaining: 2.94s
177:	learn: 5911.6604278	test: 6079.1247318	best: 6079.1247318 (177)	total: 7.19s	remaining: 2.91s
178:	learn: 5906.8344047	test: 6075.8349625	best: 6075.8349625 (178)	total: 7.24s	remaining: 2.87s
179:	learn: 5903.4390207	test: 6071.8430701	best: 6071.8430701 (179)	total: 7.29s	remaining: 2.83s
180:	learn: 5898.4196684	test: 6068.1979075	best: 6068.1979075 (180)	total: 7.34s	remaining: 2.8s
181:	learn:

6:	learn: 10962.4281835	test: 11063.3349162	best: 11063.3349162 (6)	total: 356ms	remaining: 12.4s
7:	learn: 10337.9136711	test: 10444.3087733	best: 10444.3087733 (7)	total: 407ms	remaining: 12.3s
8:	learn: 9789.2533638	test: 9901.7993864	best: 9901.7993864 (8)	total: 454ms	remaining: 12.2s
9:	learn: 9319.8935028	test: 9430.3219030	best: 9430.3219030 (9)	total: 499ms	remaining: 12s
10:	learn: 8899.1803519	test: 9018.3184622	best: 9018.3184622 (10)	total: 542ms	remaining: 11.8s
11:	learn: 8544.5132543	test: 8662.6773788	best: 8662.6773788 (11)	total: 590ms	remaining: 11.7s
12:	learn: 8239.7103333	test: 8360.5704953	best: 8360.5704953 (12)	total: 630ms	remaining: 11.5s
13:	learn: 7980.1384417	test: 8097.8585845	best: 8097.8585845 (13)	total: 669ms	remaining: 11.3s
14:	learn: 7743.8307373	test: 7864.7928817	best: 7864.7928817 (14)	total: 710ms	remaining: 11.1s
15:	learn: 7549.3810562	test: 7673.9919246	best: 7673.9919246 (15)	total: 757ms	remaining: 11.1s
16:	learn: 7375.8757080	test: 7498

93:	learn: 5544.3612153	test: 5748.9470446	best: 5748.9470446 (93)	total: 4.58s	remaining: 7.6s
94:	learn: 5532.0513003	test: 5740.8931495	best: 5740.8931495 (94)	total: 4.63s	remaining: 7.55s
95:	learn: 5518.0171748	test: 5730.8034102	best: 5730.8034102 (95)	total: 4.68s	remaining: 7.51s
96:	learn: 5503.2188945	test: 5720.5626240	best: 5720.5626240 (96)	total: 4.73s	remaining: 7.46s
97:	learn: 5493.5282962	test: 5710.5023980	best: 5710.5023980 (97)	total: 4.79s	remaining: 7.42s
98:	learn: 5484.1207288	test: 5704.2639550	best: 5704.2639550 (98)	total: 4.83s	remaining: 7.37s
99:	learn: 5478.1326971	test: 5698.8505430	best: 5698.8505430 (99)	total: 4.88s	remaining: 7.33s
100:	learn: 5466.2162459	test: 5690.3458857	best: 5690.3458857 (100)	total: 4.92s	remaining: 7.26s
101:	learn: 5453.0215462	test: 5683.9493798	best: 5683.9493798 (101)	total: 4.97s	remaining: 7.21s
102:	learn: 5445.7468513	test: 5677.3418122	best: 5677.3418122 (102)	total: 5.02s	remaining: 7.17s
103:	learn: 5438.4105097	

178:	learn: 4997.4207750	test: 5374.9175689	best: 5374.9175689 (178)	total: 7.38s	remaining: 2.93s
179:	learn: 4989.8653052	test: 5372.4059720	best: 5372.4059720 (179)	total: 7.44s	remaining: 2.89s
180:	learn: 4987.2250215	test: 5370.8065316	best: 5370.8065316 (180)	total: 7.49s	remaining: 2.85s
181:	learn: 4983.2007269	test: 5369.4023926	best: 5369.4023926 (181)	total: 7.54s	remaining: 2.82s
182:	learn: 4981.6649862	test: 5369.0739956	best: 5369.0739956 (182)	total: 7.59s	remaining: 2.78s
183:	learn: 4977.7079098	test: 5366.1999840	best: 5366.1999840 (183)	total: 7.64s	remaining: 2.74s
184:	learn: 4973.5712450	test: 5363.2630785	best: 5363.2630785 (184)	total: 7.69s	remaining: 2.7s
185:	learn: 4969.8905556	test: 5361.9631128	best: 5361.9631128 (185)	total: 7.74s	remaining: 2.66s
186:	learn: 4965.4398947	test: 5359.1725996	best: 5359.1725996 (186)	total: 7.79s	remaining: 2.63s
187:	learn: 4962.1249483	test: 5355.9938692	best: 5355.9938692 (187)	total: 7.85s	remaining: 2.59s
188:	learn:

10:	learn: 14187.6200437	test: 14275.2038855	best: 14275.2038855 (10)	total: 483ms	remaining: 10.5s
11:	learn: 13881.0236618	test: 13968.3308258	best: 13968.3308258 (11)	total: 535ms	remaining: 10.6s
12:	learn: 13584.3116724	test: 13671.5941780	best: 13671.5941780 (12)	total: 590ms	remaining: 10.8s
13:	learn: 13299.6983086	test: 13389.3699729	best: 13389.3699729 (13)	total: 641ms	remaining: 10.8s
14:	learn: 13026.3897218	test: 13114.8997130	best: 13114.8997130 (14)	total: 699ms	remaining: 11s
15:	learn: 12764.5800895	test: 12853.7763495	best: 12853.7763495 (15)	total: 760ms	remaining: 11.1s
16:	learn: 12511.8419521	test: 12601.6777684	best: 12601.6777684 (16)	total: 817ms	remaining: 11.2s
17:	learn: 12260.4797590	test: 12351.3682706	best: 12351.3682706 (17)	total: 869ms	remaining: 11.2s
18:	learn: 12025.0270148	test: 12116.7171164	best: 12116.7171164 (18)	total: 914ms	remaining: 11.1s
19:	learn: 11797.6401689	test: 11889.2462165	best: 11889.2462165 (19)	total: 960ms	remaining: 11s
20:	

97:	learn: 6421.5245969	test: 6550.9446757	best: 6550.9446757 (97)	total: 4.45s	remaining: 6.89s
98:	learn: 6410.1420761	test: 6540.8463772	best: 6540.8463772 (98)	total: 4.49s	remaining: 6.84s
99:	learn: 6399.2496549	test: 6528.9753622	best: 6528.9753622 (99)	total: 4.53s	remaining: 6.79s
100:	learn: 6388.5883768	test: 6518.6193590	best: 6518.6193590 (100)	total: 4.57s	remaining: 6.74s
101:	learn: 6377.6010265	test: 6507.6886882	best: 6507.6886882 (101)	total: 4.61s	remaining: 6.68s
102:	learn: 6367.6418026	test: 6496.6034026	best: 6496.6034026 (102)	total: 4.65s	remaining: 6.64s
103:	learn: 6358.9149823	test: 6487.4695423	best: 6487.4695423 (103)	total: 4.69s	remaining: 6.59s
104:	learn: 6350.3668066	test: 6477.9953469	best: 6477.9953469 (104)	total: 4.73s	remaining: 6.53s
105:	learn: 6340.7929099	test: 6467.9802481	best: 6467.9802481 (105)	total: 4.77s	remaining: 6.48s
106:	learn: 6330.3680568	test: 6458.8396470	best: 6458.8396470 (106)	total: 4.81s	remaining: 6.43s
107:	learn: 6321

182:	learn: 5954.4554749	test: 6090.8380183	best: 6090.8380183 (182)	total: 8.49s	remaining: 3.11s
183:	learn: 5951.1415189	test: 6088.7690965	best: 6088.7690965 (183)	total: 8.53s	remaining: 3.06s
184:	learn: 5948.6174817	test: 6086.0481982	best: 6086.0481982 (184)	total: 8.57s	remaining: 3.01s
185:	learn: 5946.3828344	test: 6084.2712605	best: 6084.2712605 (185)	total: 8.61s	remaining: 2.96s
186:	learn: 5942.9533020	test: 6081.5691225	best: 6081.5691225 (186)	total: 8.65s	remaining: 2.92s
187:	learn: 5939.0500833	test: 6077.5153853	best: 6077.5153853 (187)	total: 8.69s	remaining: 2.87s
188:	learn: 5935.5456279	test: 6074.1804642	best: 6074.1804642 (188)	total: 8.73s	remaining: 2.82s
189:	learn: 5931.7903558	test: 6071.5964515	best: 6071.5964515 (189)	total: 8.77s	remaining: 2.77s
190:	learn: 5926.2311387	test: 6067.4813337	best: 6067.4813337 (190)	total: 8.81s	remaining: 2.72s
191:	learn: 5923.6031933	test: 6065.1334461	best: 6065.1334461 (191)	total: 8.85s	remaining: 2.67s
192:	learn

22:	learn: 6729.8055825	test: 6862.7847081	best: 6862.7847081 (22)	total: 610ms	remaining: 6.02s
23:	learn: 6666.3896841	test: 6800.8662024	best: 6800.8662024 (23)	total: 640ms	remaining: 6.02s
24:	learn: 6613.0622090	test: 6744.1294299	best: 6744.1294299 (24)	total: 668ms	remaining: 6.01s
25:	learn: 6561.6203736	test: 6690.5633921	best: 6690.5633921 (25)	total: 699ms	remaining: 6.02s
26:	learn: 6506.9794895	test: 6641.1597068	best: 6641.1597068 (26)	total: 732ms	remaining: 6.04s
27:	learn: 6467.4871394	test: 6598.9140245	best: 6598.9140245 (27)	total: 761ms	remaining: 6.03s
28:	learn: 6425.7402308	test: 6558.6887078	best: 6558.6887078 (28)	total: 784ms	remaining: 5.98s
29:	learn: 6393.6479971	test: 6523.4310698	best: 6523.4310698 (29)	total: 807ms	remaining: 5.92s
30:	learn: 6367.2606934	test: 6494.9410792	best: 6494.9410792 (30)	total: 831ms	remaining: 5.87s
31:	learn: 6336.6104929	test: 6462.6040049	best: 6462.6040049 (31)	total: 853ms	remaining: 5.81s
32:	learn: 6301.6890839	test: 

108:	learn: 5496.6050371	test: 5715.8411998	best: 5715.8411998 (108)	total: 5s	remaining: 6.46s
109:	learn: 5487.2460285	test: 5710.1231144	best: 5710.1231144 (109)	total: 5.04s	remaining: 6.42s
110:	learn: 5479.2781585	test: 5704.1652492	best: 5704.1652492 (110)	total: 5.09s	remaining: 6.37s
111:	learn: 5473.6799919	test: 5697.2290502	best: 5697.2290502 (111)	total: 5.14s	remaining: 6.33s
112:	learn: 5468.6886891	test: 5692.4246964	best: 5692.4246964 (112)	total: 5.2s	remaining: 6.31s
113:	learn: 5464.6569766	test: 5687.4778758	best: 5687.4778758 (113)	total: 5.26s	remaining: 6.28s
114:	learn: 5456.1467407	test: 5681.6868793	best: 5681.6868793 (114)	total: 5.33s	remaining: 6.26s
115:	learn: 5449.6887738	test: 5679.7371874	best: 5679.7371874 (115)	total: 5.38s	remaining: 6.21s
116:	learn: 5442.6047641	test: 5676.8628007	best: 5676.8628007 (116)	total: 5.44s	remaining: 6.18s
117:	learn: 5437.7536245	test: 5671.7561800	best: 5671.7561800 (117)	total: 5.49s	remaining: 6.14s
118:	learn: 54

191:	learn: 5064.4339507	test: 5424.5125726	best: 5424.5125726 (191)	total: 9.51s	remaining: 2.87s
192:	learn: 5061.9631669	test: 5423.4058107	best: 5423.4058107 (192)	total: 9.56s	remaining: 2.82s
193:	learn: 5057.9359479	test: 5421.7792085	best: 5421.7792085 (193)	total: 9.61s	remaining: 2.77s
194:	learn: 5054.7537133	test: 5420.1740063	best: 5420.1740063 (194)	total: 9.65s	remaining: 2.72s
195:	learn: 5051.9983963	test: 5418.9757842	best: 5418.9757842 (195)	total: 9.69s	remaining: 2.67s
196:	learn: 5047.0731953	test: 5416.1495317	best: 5416.1495317 (196)	total: 9.74s	remaining: 2.62s
197:	learn: 5042.2017341	test: 5414.3058104	best: 5414.3058104 (197)	total: 9.78s	remaining: 2.57s
198:	learn: 5036.8955441	test: 5412.4593181	best: 5412.4593181 (198)	total: 9.83s	remaining: 2.52s
199:	learn: 5033.6317084	test: 5410.8540045	best: 5410.8540045 (199)	total: 9.88s	remaining: 2.47s
200:	learn: 5028.9944287	test: 5407.9522180	best: 5407.9522180 (200)	total: 9.93s	remaining: 2.42s
201:	learn

23:	learn: 6420.3858061	test: 6900.3748706	best: 6900.3748706 (23)	total: 1.07s	remaining: 10.1s
24:	learn: 6362.1713560	test: 6845.8969836	best: 6845.8969836 (24)	total: 1.12s	remaining: 10s
25:	learn: 6302.9191887	test: 6797.8892962	best: 6797.8892962 (25)	total: 1.16s	remaining: 9.96s
26:	learn: 6250.3898300	test: 6748.1730373	best: 6748.1730373 (26)	total: 1.19s	remaining: 9.86s
27:	learn: 6203.3752416	test: 6703.5047156	best: 6703.5047156 (27)	total: 1.24s	remaining: 9.79s
28:	learn: 6163.1382201	test: 6668.7488836	best: 6668.7488836 (28)	total: 1.28s	remaining: 9.75s
29:	learn: 6125.8164026	test: 6633.3856277	best: 6633.3856277 (29)	total: 1.33s	remaining: 9.75s
30:	learn: 6097.9296063	test: 6609.3515769	best: 6609.3515769 (30)	total: 1.38s	remaining: 9.74s
31:	learn: 6064.7585033	test: 6581.6509327	best: 6581.6509327 (31)	total: 1.43s	remaining: 9.71s
32:	learn: 6038.5755596	test: 6559.4424877	best: 6559.4424877 (32)	total: 1.47s	remaining: 9.69s
33:	learn: 6010.2690551	test: 65

108:	learn: 5079.9710733	test: 5888.3718556	best: 5888.3718556 (108)	total: 4.7s	remaining: 6.08s
109:	learn: 5066.9237710	test: 5880.2424325	best: 5880.2424325 (109)	total: 4.75s	remaining: 6.04s
110:	learn: 5062.7678381	test: 5876.8380184	best: 5876.8380184 (110)	total: 4.79s	remaining: 5.99s
111:	learn: 5052.2209718	test: 5873.4958136	best: 5873.4958136 (111)	total: 4.82s	remaining: 5.94s
112:	learn: 5043.4531412	test: 5868.6827296	best: 5868.6827296 (112)	total: 4.86s	remaining: 5.89s
113:	learn: 5039.6700954	test: 5865.2308266	best: 5865.2308266 (113)	total: 4.9s	remaining: 5.85s
114:	learn: 5034.9398394	test: 5858.0253317	best: 5858.0253317 (114)	total: 4.95s	remaining: 5.81s
115:	learn: 5029.5066974	test: 5853.5859067	best: 5853.5859067 (115)	total: 4.99s	remaining: 5.76s
116:	learn: 5019.8696547	test: 5850.1047209	best: 5850.1047209 (116)	total: 5.03s	remaining: 5.72s
117:	learn: 5016.3545033	test: 5846.5924315	best: 5846.5924315 (117)	total: 5.07s	remaining: 5.67s
118:	learn: 

199:	learn: 4532.8131868	test: 5645.5842746	best: 5645.5842746 (199)	total: 7.59s	remaining: 1.9s
200:	learn: 4526.0037282	test: 5642.8367056	best: 5642.8367056 (200)	total: 7.62s	remaining: 1.86s
201:	learn: 4520.9694066	test: 5641.5723436	best: 5641.5723436 (201)	total: 7.64s	remaining: 1.81s
202:	learn: 4516.3880564	test: 5640.3715927	best: 5640.3715927 (202)	total: 7.68s	remaining: 1.78s
203:	learn: 4510.2851163	test: 5638.5823226	best: 5638.5823226 (203)	total: 7.73s	remaining: 1.74s
204:	learn: 4505.6315507	test: 5636.8189399	best: 5636.8189399 (204)	total: 7.77s	remaining: 1.71s
205:	learn: 4503.3457089	test: 5635.0785968	best: 5635.0785968 (205)	total: 7.82s	remaining: 1.67s
206:	learn: 4498.4009324	test: 5634.3840508	best: 5634.3840508 (206)	total: 7.86s	remaining: 1.63s
207:	learn: 4494.7115798	test: 5633.3690584	best: 5633.3690584 (207)	total: 7.91s	remaining: 1.6s
208:	learn: 4486.6534505	test: 5630.4668196	best: 5630.4668196 (208)	total: 7.96s	remaining: 1.56s
209:	learn: 

33:	learn: 6206.1475876	test: 6155.7147316	best: 6155.7147316 (33)	total: 1.5s	remaining: 9.52s
34:	learn: 6183.8165546	test: 6134.0345682	best: 6134.0345682 (34)	total: 1.54s	remaining: 9.48s
35:	learn: 6164.5825814	test: 6117.9884746	best: 6117.9884746 (35)	total: 1.58s	remaining: 9.39s
36:	learn: 6137.1088512	test: 6094.8270445	best: 6094.8270445 (36)	total: 1.62s	remaining: 9.33s
37:	learn: 6117.1952627	test: 6076.6628439	best: 6076.6628439 (37)	total: 1.66s	remaining: 9.25s
38:	learn: 6097.0960121	test: 6058.4956123	best: 6058.4956123 (38)	total: 1.7s	remaining: 9.19s
39:	learn: 6069.5321439	test: 6039.0861738	best: 6039.0861738 (39)	total: 1.74s	remaining: 9.16s
40:	learn: 6054.8257611	test: 6024.0697238	best: 6024.0697238 (40)	total: 1.78s	remaining: 9.09s
41:	learn: 6034.8369783	test: 6007.7167496	best: 6007.7167496 (41)	total: 1.82s	remaining: 9.03s
42:	learn: 6018.9147195	test: 5992.3876218	best: 5992.3876218 (42)	total: 1.86s	remaining: 8.95s
43:	learn: 6004.7175309	test: 59

121:	learn: 5211.8696750	test: 5460.8054968	best: 5460.8054968 (121)	total: 4.75s	remaining: 4.99s
122:	learn: 5205.1453235	test: 5456.8824866	best: 5456.8824866 (122)	total: 4.77s	remaining: 4.93s
123:	learn: 5202.3616069	test: 5454.6474017	best: 5454.6474017 (123)	total: 4.79s	remaining: 4.87s
124:	learn: 5193.5181663	test: 5451.0265467	best: 5451.0265467 (124)	total: 4.81s	remaining: 4.81s
125:	learn: 5186.0031674	test: 5447.8911891	best: 5447.8911891 (125)	total: 4.83s	remaining: 4.75s
126:	learn: 5178.5485912	test: 5443.7505965	best: 5443.7505965 (126)	total: 4.85s	remaining: 4.7s
127:	learn: 5169.3418174	test: 5441.6265343	best: 5441.6265343 (127)	total: 4.88s	remaining: 4.65s
128:	learn: 5159.8259305	test: 5437.7241993	best: 5437.7241993 (128)	total: 4.9s	remaining: 4.59s
129:	learn: 5150.0579309	test: 5431.5222944	best: 5431.5222944 (129)	total: 4.92s	remaining: 4.54s
130:	learn: 5144.7614261	test: 5426.0044909	best: 5426.0044909 (130)	total: 4.94s	remaining: 4.49s
131:	learn: 

205:	learn: 4708.0813393	test: 5241.1071279	best: 5241.1071279 (205)	total: 7.36s	remaining: 1.57s
206:	learn: 4704.9843193	test: 5239.3010443	best: 5239.3010443 (206)	total: 7.4s	remaining: 1.54s
207:	learn: 4697.8344966	test: 5237.1359804	best: 5237.1359804 (207)	total: 7.44s	remaining: 1.5s
208:	learn: 4691.7091980	test: 5234.9616711	best: 5234.9616711 (208)	total: 7.48s	remaining: 1.47s
209:	learn: 4687.6224928	test: 5231.6948019	best: 5231.6948019 (209)	total: 7.51s	remaining: 1.43s
210:	learn: 4683.6569826	test: 5230.3402453	best: 5230.3402453 (210)	total: 7.55s	remaining: 1.4s
211:	learn: 4680.1689411	test: 5228.8168167	best: 5228.8168167 (211)	total: 7.6s	remaining: 1.36s
212:	learn: 4673.7535400	test: 5227.1309643	best: 5227.1309643 (212)	total: 7.64s	remaining: 1.33s
213:	learn: 4670.0753074	test: 5225.9025252	best: 5225.9025252 (213)	total: 7.68s	remaining: 1.29s
214:	learn: 4665.9597410	test: 5224.2280116	best: 5224.2280116 (214)	total: 7.71s	remaining: 1.25s
215:	learn: 46

41:	learn: 5986.1276260	test: 6150.6444401	best: 6150.6444401 (41)	total: 1.78s	remaining: 8.84s
42:	learn: 5966.8726740	test: 6137.3371626	best: 6137.3371626 (42)	total: 1.82s	remaining: 8.79s
43:	learn: 5948.0479559	test: 6124.0935881	best: 6124.0935881 (43)	total: 1.87s	remaining: 8.74s
44:	learn: 5936.5263411	test: 6111.5516597	best: 6111.5516597 (44)	total: 1.9s	remaining: 8.68s
45:	learn: 5924.9952941	test: 6099.4824357	best: 6099.4824357 (45)	total: 1.94s	remaining: 8.6s
46:	learn: 5913.4113015	test: 6089.6168452	best: 6089.6168452 (46)	total: 1.98s	remaining: 8.54s
47:	learn: 5898.0355948	test: 6079.3188316	best: 6079.3188316 (47)	total: 2.02s	remaining: 8.49s
48:	learn: 5884.1467608	test: 6069.5859854	best: 6069.5859854 (48)	total: 2.06s	remaining: 8.43s
49:	learn: 5871.2581818	test: 6056.2285075	best: 6056.2285075 (49)	total: 2.09s	remaining: 8.37s
50:	learn: 5858.3817747	test: 6046.4145903	best: 6046.4145903 (50)	total: 2.13s	remaining: 8.31s
51:	learn: 5834.5533035	test: 60

127:	learn: 5138.6631716	test: 5569.2439687	best: 5569.2439687 (127)	total: 4.16s	remaining: 3.96s
128:	learn: 5129.7676117	test: 5564.3348918	best: 5564.3348918 (128)	total: 4.18s	remaining: 3.92s
129:	learn: 5122.3240296	test: 5560.6603647	best: 5560.6603647 (129)	total: 4.2s	remaining: 3.88s
130:	learn: 5112.0154732	test: 5556.6232567	best: 5556.6232567 (130)	total: 4.22s	remaining: 3.84s
131:	learn: 5103.8365465	test: 5554.1084082	best: 5554.1084082 (131)	total: 4.26s	remaining: 3.81s
132:	learn: 5098.0803025	test: 5548.8302611	best: 5548.8302611 (132)	total: 4.3s	remaining: 3.78s
133:	learn: 5090.3465830	test: 5545.5614532	best: 5545.5614532 (133)	total: 4.33s	remaining: 3.75s
134:	learn: 5081.4048063	test: 5538.5602669	best: 5538.5602669 (134)	total: 4.37s	remaining: 3.73s
135:	learn: 5074.1017055	test: 5535.6280398	best: 5535.6280398 (135)	total: 4.42s	remaining: 3.7s
136:	learn: 5068.1321430	test: 5532.1213103	best: 5532.1213103 (136)	total: 4.45s	remaining: 3.67s
137:	learn: 5

210:	learn: 4651.4044725	test: 5339.2899243	best: 5339.2899243 (210)	total: 7.39s	remaining: 1.37s
211:	learn: 4643.8396418	test: 5337.2378313	best: 5337.2378313 (211)	total: 7.44s	remaining: 1.33s
212:	learn: 4638.9131002	test: 5335.6908232	best: 5335.6908232 (212)	total: 7.47s	remaining: 1.3s
213:	learn: 4635.1228792	test: 5334.4233905	best: 5334.4233905 (213)	total: 7.51s	remaining: 1.26s
214:	learn: 4631.4892604	test: 5333.6493321	best: 5333.6493321 (214)	total: 7.55s	remaining: 1.23s
215:	learn: 4627.9156862	test: 5331.5864396	best: 5331.5864396 (215)	total: 7.59s	remaining: 1.19s
216:	learn: 4625.2573167	test: 5329.6214298	best: 5329.6214298 (216)	total: 7.63s	remaining: 1.16s
217:	learn: 4622.4732304	test: 5327.4265044	best: 5327.4265044 (217)	total: 7.67s	remaining: 1.13s
218:	learn: 4617.4832918	test: 5325.4958808	best: 5325.4958808 (218)	total: 7.71s	remaining: 1.09s
219:	learn: 4612.7466284	test: 5323.7244157	best: 5323.7244157 (219)	total: 7.75s	remaining: 1.06s
220:	learn:

`Hyper-parameter Tuning`

* r2_score =  0.785446636046637

![R2_score](img.png)

![Overall Model Performance](op2.png)

### Overall performance by the different models,
> **Random Forest Regression** gives the **highest R2_score = 0.818496003833452**

df_check = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})
df_check = df_check.head(25)
#round(df_check,2)
df_check.plot(kind='bar',figsize=(10,5))
plt.grid(which='major', linestyle='-', linewidth='0.5', color='green')
plt.title('Performance of XGBRegressor')
plt.ylabel('Mean Squared Log Error')
plt.savefig('XGBRegressor-Performance.jpg')
plt.show()

`
#Ploting feature importance graph
importances=model.feature_importances_
features=X_train.columns

# Set the style
plt.style.use('fivethirtyeight')
# list of x locations for plotting
x_values = list(range(len(importances)))
# Make a bar chart
plt.figure(figsize=(6,6))
plt.bar(x_values, importances, orientation = ('vertical'))
# Tick labels for x axis
plt.xticks(x_values, features, rotation=(90))
# Axis labels and title
plt.ylabel('Importance'); 
plt.xlabel('Variable/Features'); 
plt.title('XGBRegressor Variables Importance')
plt.tight_layout()
plt.savefig('XGBRegressor-Variables-Importances.jpg',dpi=600)
plt.show();`